{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "8umqo5XJ6sFs"
   },
   "source": [
    "### Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "executionInfo": {
     "elapsed": 12,
     "status": "ok",
     "timestamp": 1710577246630,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "hOqUByRp7OYH"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "from combat.pycombat import pycombat\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "def load_json(json_path):\n",
    "    \"\"\"\n",
    "    Load json data as a dictionary\n",
    "    \"\"\"\n",
    "    with open(json_path) as f:\n",
    "        file = f.read()\n",
    "    json_data = json.loads(file)\n",
    "\n",
    "    return json_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "RHpljGrL7Tz9"
   },
   "source": [
    "### Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['AFF_PM_to_ALIAS', 'ENTREZID_to_ALIAS'])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load Annotation\n",
    "Affimetrix_map = load_json(\"data2/Affymetrix_map.json\")\n",
    "Affimetrix_map.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene expression data\n",
    "GSE152004_norm_data = pd.read_csv('data1/GSE152004_norm_data.csv')\n",
    "GSE67472_norm_data = pd.read_csv('data1/GSE67472_norm_data.csv')\n",
    "GSE201955_norm_data = pd.read_csv('data1/GSE201955_norm_data.csv')\n",
    "GSE58434_norm_data = pd.read_csv('data1/GSE58434_norm_data.csv')\n",
    "GSE69683_norm_data = pd.read_csv('data1/GSE69683_norm_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata\n",
    "GSE152004_metadata = pd.read_csv('data1/GSE152004_metadata.csv')\n",
    "GSE152004_metadata.index = GSE152004_metadata[\"Unnamed: 0\"]\n",
    "GSE152004_metadata.drop(\"Unnamed: 0\", axis = 1, inplace = True)\n",
    "\n",
    "GSE67472_metadata = pd.read_csv('data1/GSE67472_metadata.csv')\n",
    "GSE67472_metadata.index = GSE67472_metadata[\"Unnamed: 0\"]\n",
    "GSE67472_metadata.drop(\"Unnamed: 0\", axis = 1, inplace = True)\n",
    "\n",
    "GSE201955_metadata = pd.read_csv('data1/GSE201955_metadata.csv')\n",
    "GSE201955_metadata.index = GSE201955_metadata[\"Unnamed: 0\"]\n",
    "GSE201955_metadata.drop(\"Unnamed: 0\", axis = 1, inplace = True)\n",
    "\n",
    "GSE58434_metadata = pd.read_csv('data1/GSE58434_metadata.csv')\n",
    "GSE58434_metadata.index = GSE58434_metadata[\"Unnamed: 0\"]\n",
    "GSE58434_metadata.drop(\"Unnamed: 0\", axis = 1, inplace = True)\n",
    "\n",
    "GSE69683_metadata = pd.read_csv('data1/GSE69683_metadata.csv')\n",
    "GSE69683_metadata.index = GSE69683_metadata[\"Unnamed: 0\"]\n",
    "GSE69683_metadata.drop(\"Unnamed: 0\", axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Limma, and WGCNA Results\n",
    "GSE152004_modules_df = pd.read_csv('data2/GSE152004_modules.csv')\n",
    "GSE152004_Limma_Results_df = pd.read_csv('data2/GSE152004_Limma_Results.csv')\n",
    "GSE152004_r_and_Pvalue_df = pd.read_csv('data2/GSE152004_r_and_Pvalue_data.csv')\n",
    "\n",
    "GSE67472_modules_df = pd.read_csv('data2/GSE67472_modules.csv')\n",
    "GSE67472_Limma_Results_df = pd.read_csv('data2/GSE67472_Limma_Results.csv')\n",
    "GSE67472_r_and_Pvalue_df = pd.read_csv('data2/GSE67472_r_and_Pvalue_data.csv')\n",
    "\n",
    "GSE201955_modules_df = pd.read_csv('data2/GSE201955_modules.csv')\n",
    "GSE201955_Limma_Results_df = pd.read_csv('data2/GSE201955_Limma_Results.csv')\n",
    "GSE201955_r_and_Pvalue_df = pd.read_csv('data2/GSE201955_r_and_Pvalue_data.csv')\n",
    "\n",
    "GSE58434_modules_df = pd.read_csv('data2/GSE58434_modules.csv')\n",
    "GSE58434_Limma_Results_df = pd.read_csv('data2/GSE58434_Limma_Results.csv')\n",
    "GSE58434_r_and_Pvalue_df = pd.read_csv('data2/GSE58434_r_and_Pvalue_data.csv')\n",
    "\n",
    "GSE69683_modules_df = pd.read_csv('data2/GSE69683_modules.csv')\n",
    "GSE69683_Limma_Results_df = pd.read_csv('data2/GSE69683_Limma_Results.csv')\n",
    "GSE69683_r_and_Pvalue_df = pd.read_csv('data2/GSE69683_r_and_Pvalue_data.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Arrange Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_id</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HR1005</th>\n",
       "      <td>12.047600</td>\n",
       "      <td>12.047600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>531.815488</td>\n",
       "      <td>75.727772</td>\n",
       "      <td>151.455543</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.047600</td>\n",
       "      <td>7380.015570</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>333.890630</td>\n",
       "      <td>165.224229</td>\n",
       "      <td>213.414629</td>\n",
       "      <td>487.067259</td>\n",
       "      <td>1485.296977</td>\n",
       "      <td>15.489771</td>\n",
       "      <td>1519.718691</td>\n",
       "      <td>2039.486579</td>\n",
       "      <td>4042.830358</td>\n",
       "      <td>1089.447261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1006</th>\n",
       "      <td>16.204622</td>\n",
       "      <td>13.645997</td>\n",
       "      <td>0.0</td>\n",
       "      <td>382.940804</td>\n",
       "      <td>23.880496</td>\n",
       "      <td>633.686008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.705750</td>\n",
       "      <td>20669.421796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>326.651064</td>\n",
       "      <td>121.961102</td>\n",
       "      <td>187.632465</td>\n",
       "      <td>508.313406</td>\n",
       "      <td>1739.864678</td>\n",
       "      <td>52.025365</td>\n",
       "      <td>1409.802114</td>\n",
       "      <td>3917.254149</td>\n",
       "      <td>4017.040506</td>\n",
       "      <td>1224.728273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1009</th>\n",
       "      <td>21.992452</td>\n",
       "      <td>12.936737</td>\n",
       "      <td>0.0</td>\n",
       "      <td>169.471249</td>\n",
       "      <td>29.754494</td>\n",
       "      <td>648.130502</td>\n",
       "      <td>1.293674</td>\n",
       "      <td>9.055716</td>\n",
       "      <td>12044.101737</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>394.570465</td>\n",
       "      <td>249.679016</td>\n",
       "      <td>204.400438</td>\n",
       "      <td>512.294768</td>\n",
       "      <td>1503.248788</td>\n",
       "      <td>67.271030</td>\n",
       "      <td>1582.162881</td>\n",
       "      <td>2668.848752</td>\n",
       "      <td>3708.962372</td>\n",
       "      <td>1353.182644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1010</th>\n",
       "      <td>9.808778</td>\n",
       "      <td>3.772607</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.532202</td>\n",
       "      <td>26.408249</td>\n",
       "      <td>705.477508</td>\n",
       "      <td>0.754521</td>\n",
       "      <td>8.299735</td>\n",
       "      <td>18126.622082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>467.048746</td>\n",
       "      <td>273.891268</td>\n",
       "      <td>202.966256</td>\n",
       "      <td>605.880683</td>\n",
       "      <td>1489.425241</td>\n",
       "      <td>51.307455</td>\n",
       "      <td>1876.494719</td>\n",
       "      <td>3338.757189</td>\n",
       "      <td>3414.963850</td>\n",
       "      <td>1251.751000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1012</th>\n",
       "      <td>11.376166</td>\n",
       "      <td>21.330311</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1083.579785</td>\n",
       "      <td>28.440414</td>\n",
       "      <td>203.348962</td>\n",
       "      <td>4.266062</td>\n",
       "      <td>5.688083</td>\n",
       "      <td>13180.710004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>462.156732</td>\n",
       "      <td>314.266578</td>\n",
       "      <td>214.725128</td>\n",
       "      <td>484.909064</td>\n",
       "      <td>1275.552581</td>\n",
       "      <td>27.018394</td>\n",
       "      <td>1504.497916</td>\n",
       "      <td>3924.777172</td>\n",
       "      <td>3431.335985</td>\n",
       "      <td>1215.827711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26485 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_id           A1BG   A1BG-AS1  A1CF          A2M    A2M-AS1       A2ML1  \\\n",
       "sample_name                                                                   \n",
       "HR1005       12.047600  12.047600   0.0   531.815488  75.727772  151.455543   \n",
       "HR1006       16.204622  13.645997   0.0   382.940804  23.880496  633.686008   \n",
       "HR1009       21.992452  12.936737   0.0   169.471249  29.754494  648.130502   \n",
       "HR1010        9.808778   3.772607   0.0   130.532202  26.408249  705.477508   \n",
       "HR1012       11.376166  21.330311   0.0  1083.579785  28.440414  203.348962   \n",
       "\n",
       "gene_id         A2MP1    A3GALT2        A4GALT  A4GNT  ...      ZWILCH  \\\n",
       "sample_name                                            ...               \n",
       "HR1005       0.000000  12.047600   7380.015570    0.0  ...  333.890630   \n",
       "HR1006       0.000000   1.705750  20669.421796    0.0  ...  326.651064   \n",
       "HR1009       1.293674   9.055716  12044.101737    0.0  ...  394.570465   \n",
       "HR1010       0.754521   8.299735  18126.622082    0.0  ...  467.048746   \n",
       "HR1012       4.266062   5.688083  13180.710004    0.0  ...  462.156732   \n",
       "\n",
       "gene_id           ZWINT        ZXDA        ZXDB         ZXDC     ZYG11A  \\\n",
       "sample_name                                                               \n",
       "HR1005       165.224229  213.414629  487.067259  1485.296977  15.489771   \n",
       "HR1006       121.961102  187.632465  508.313406  1739.864678  52.025365   \n",
       "HR1009       249.679016  204.400438  512.294768  1503.248788  67.271030   \n",
       "HR1010       273.891268  202.966256  605.880683  1489.425241  51.307455   \n",
       "HR1012       314.266578  214.725128  484.909064  1275.552581  27.018394   \n",
       "\n",
       "gene_id           ZYG11B          ZYX        ZZEF1         ZZZ3  \n",
       "sample_name                                                      \n",
       "HR1005       1519.718691  2039.486579  4042.830358  1089.447261  \n",
       "HR1006       1409.802114  3917.254149  4017.040506  1224.728273  \n",
       "HR1009       1582.162881  2668.848752  3708.962372  1353.182644  \n",
       "HR1010       1876.494719  3338.757189  3414.963850  1251.751000  \n",
       "HR1012       1504.497916  3924.777172  3431.335985  1215.827711  \n",
       "\n",
       "[5 rows x 26485 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE152004_norm_data.index = GSE152004_norm_data.gene_id\n",
    "GSE152004_norm_data.drop(\"gene_id\", axis = 1, inplace = True)\n",
    "GSE152004_norm_data = GSE152004_norm_data.T\n",
    "GSE152004_norm_data.index.name = \"sample_name\"\n",
    "GSE152004_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_id</th>\n",
       "      <th>100009676_at</th>\n",
       "      <th>10000_at</th>\n",
       "      <th>10001_at</th>\n",
       "      <th>10002_at</th>\n",
       "      <th>10003_at</th>\n",
       "      <th>100048912_at</th>\n",
       "      <th>100049716_at</th>\n",
       "      <th>10004_at</th>\n",
       "      <th>10005_at</th>\n",
       "      <th>10006_at</th>\n",
       "      <th>...</th>\n",
       "      <th>9989_at</th>\n",
       "      <th>998_at</th>\n",
       "      <th>9990_at</th>\n",
       "      <th>9991_at</th>\n",
       "      <th>9992_at</th>\n",
       "      <th>9993_at</th>\n",
       "      <th>9994_at</th>\n",
       "      <th>9997_at</th>\n",
       "      <th>999_at</th>\n",
       "      <th>9_at</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 01</th>\n",
       "      <td>5.329072</td>\n",
       "      <td>4.888653</td>\n",
       "      <td>6.717174</td>\n",
       "      <td>4.903333</td>\n",
       "      <td>4.445861</td>\n",
       "      <td>4.808544</td>\n",
       "      <td>6.854104</td>\n",
       "      <td>5.103949</td>\n",
       "      <td>5.927904</td>\n",
       "      <td>9.579241</td>\n",
       "      <td>...</td>\n",
       "      <td>10.117098</td>\n",
       "      <td>7.123103</td>\n",
       "      <td>6.346994</td>\n",
       "      <td>8.523364</td>\n",
       "      <td>5.859315</td>\n",
       "      <td>7.303053</td>\n",
       "      <td>5.222605</td>\n",
       "      <td>9.679513</td>\n",
       "      <td>8.545118</td>\n",
       "      <td>9.604065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 02</th>\n",
       "      <td>5.995749</td>\n",
       "      <td>4.980210</td>\n",
       "      <td>6.641973</td>\n",
       "      <td>4.991354</td>\n",
       "      <td>4.354581</td>\n",
       "      <td>4.639297</td>\n",
       "      <td>6.299639</td>\n",
       "      <td>5.765387</td>\n",
       "      <td>6.345293</td>\n",
       "      <td>8.970563</td>\n",
       "      <td>...</td>\n",
       "      <td>9.711283</td>\n",
       "      <td>7.067920</td>\n",
       "      <td>6.315463</td>\n",
       "      <td>8.517702</td>\n",
       "      <td>5.299672</td>\n",
       "      <td>7.654083</td>\n",
       "      <td>4.410016</td>\n",
       "      <td>8.691703</td>\n",
       "      <td>9.025900</td>\n",
       "      <td>9.388568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 03</th>\n",
       "      <td>5.702682</td>\n",
       "      <td>5.371424</td>\n",
       "      <td>7.234998</td>\n",
       "      <td>5.398429</td>\n",
       "      <td>4.379199</td>\n",
       "      <td>4.752368</td>\n",
       "      <td>6.337287</td>\n",
       "      <td>5.251048</td>\n",
       "      <td>5.843571</td>\n",
       "      <td>9.537878</td>\n",
       "      <td>...</td>\n",
       "      <td>9.499548</td>\n",
       "      <td>7.222755</td>\n",
       "      <td>6.033887</td>\n",
       "      <td>8.550329</td>\n",
       "      <td>5.482077</td>\n",
       "      <td>7.495711</td>\n",
       "      <td>4.537349</td>\n",
       "      <td>7.265700</td>\n",
       "      <td>8.853830</td>\n",
       "      <td>9.350149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, healthy sample 01</th>\n",
       "      <td>5.583717</td>\n",
       "      <td>4.834860</td>\n",
       "      <td>7.108063</td>\n",
       "      <td>5.141136</td>\n",
       "      <td>4.159465</td>\n",
       "      <td>5.291560</td>\n",
       "      <td>6.255015</td>\n",
       "      <td>5.151088</td>\n",
       "      <td>5.987954</td>\n",
       "      <td>9.131043</td>\n",
       "      <td>...</td>\n",
       "      <td>9.964046</td>\n",
       "      <td>7.217349</td>\n",
       "      <td>6.062817</td>\n",
       "      <td>8.621510</td>\n",
       "      <td>5.015522</td>\n",
       "      <td>7.563695</td>\n",
       "      <td>4.613191</td>\n",
       "      <td>9.566222</td>\n",
       "      <td>8.849365</td>\n",
       "      <td>9.172806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 04</th>\n",
       "      <td>5.743467</td>\n",
       "      <td>4.736444</td>\n",
       "      <td>7.148278</td>\n",
       "      <td>4.757029</td>\n",
       "      <td>4.754785</td>\n",
       "      <td>4.487500</td>\n",
       "      <td>6.468263</td>\n",
       "      <td>5.425693</td>\n",
       "      <td>6.053634</td>\n",
       "      <td>9.198382</td>\n",
       "      <td>...</td>\n",
       "      <td>10.091075</td>\n",
       "      <td>6.835439</td>\n",
       "      <td>6.413736</td>\n",
       "      <td>8.540482</td>\n",
       "      <td>5.165816</td>\n",
       "      <td>7.486327</td>\n",
       "      <td>4.374204</td>\n",
       "      <td>9.074473</td>\n",
       "      <td>8.714443</td>\n",
       "      <td>9.346384</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19008 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_id                               100009676_at  10000_at  10001_at  \\\n",
       "sample_name                                                              \n",
       "Airway epithelium, asthma sample 01       5.329072  4.888653  6.717174   \n",
       "Airway epithelium, asthma sample 02       5.995749  4.980210  6.641973   \n",
       "Airway epithelium, asthma sample 03       5.702682  5.371424  7.234998   \n",
       "Airway epithelium, healthy sample 01      5.583717  4.834860  7.108063   \n",
       "Airway epithelium, asthma sample 04       5.743467  4.736444  7.148278   \n",
       "\n",
       "gene_id                               10002_at  10003_at  100048912_at  \\\n",
       "sample_name                                                              \n",
       "Airway epithelium, asthma sample 01   4.903333  4.445861      4.808544   \n",
       "Airway epithelium, asthma sample 02   4.991354  4.354581      4.639297   \n",
       "Airway epithelium, asthma sample 03   5.398429  4.379199      4.752368   \n",
       "Airway epithelium, healthy sample 01  5.141136  4.159465      5.291560   \n",
       "Airway epithelium, asthma sample 04   4.757029  4.754785      4.487500   \n",
       "\n",
       "gene_id                               100049716_at  10004_at  10005_at  \\\n",
       "sample_name                                                              \n",
       "Airway epithelium, asthma sample 01       6.854104  5.103949  5.927904   \n",
       "Airway epithelium, asthma sample 02       6.299639  5.765387  6.345293   \n",
       "Airway epithelium, asthma sample 03       6.337287  5.251048  5.843571   \n",
       "Airway epithelium, healthy sample 01      6.255015  5.151088  5.987954   \n",
       "Airway epithelium, asthma sample 04       6.468263  5.425693  6.053634   \n",
       "\n",
       "gene_id                               10006_at  ...    9989_at    998_at  \\\n",
       "sample_name                                     ...                        \n",
       "Airway epithelium, asthma sample 01   9.579241  ...  10.117098  7.123103   \n",
       "Airway epithelium, asthma sample 02   8.970563  ...   9.711283  7.067920   \n",
       "Airway epithelium, asthma sample 03   9.537878  ...   9.499548  7.222755   \n",
       "Airway epithelium, healthy sample 01  9.131043  ...   9.964046  7.217349   \n",
       "Airway epithelium, asthma sample 04   9.198382  ...  10.091075  6.835439   \n",
       "\n",
       "gene_id                                9990_at   9991_at   9992_at   9993_at  \\\n",
       "sample_name                                                                    \n",
       "Airway epithelium, asthma sample 01   6.346994  8.523364  5.859315  7.303053   \n",
       "Airway epithelium, asthma sample 02   6.315463  8.517702  5.299672  7.654083   \n",
       "Airway epithelium, asthma sample 03   6.033887  8.550329  5.482077  7.495711   \n",
       "Airway epithelium, healthy sample 01  6.062817  8.621510  5.015522  7.563695   \n",
       "Airway epithelium, asthma sample 04   6.413736  8.540482  5.165816  7.486327   \n",
       "\n",
       "gene_id                                9994_at   9997_at    999_at      9_at  \n",
       "sample_name                                                                   \n",
       "Airway epithelium, asthma sample 01   5.222605  9.679513  8.545118  9.604065  \n",
       "Airway epithelium, asthma sample 02   4.410016  8.691703  9.025900  9.388568  \n",
       "Airway epithelium, asthma sample 03   4.537349  7.265700  8.853830  9.350149  \n",
       "Airway epithelium, healthy sample 01  4.613191  9.566222  8.849365  9.172806  \n",
       "Airway epithelium, asthma sample 04   4.374204  9.074473  8.714443  9.346384  \n",
       "\n",
       "[5 rows x 19008 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE67472_norm_data.index = GSE67472_norm_data.gene_id\n",
    "GSE67472_norm_data.drop(\"gene_id\", axis = 1, inplace = True)\n",
    "GSE67472_norm_data = GSE67472_norm_data.T\n",
    "GSE67472_norm_data.index.name = \"sample_name\"\n",
    "GSE67472_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_id</th>\n",
       "      <th>DNAJC11</th>\n",
       "      <th>CDK11A</th>\n",
       "      <th>NADK</th>\n",
       "      <th>MASP2</th>\n",
       "      <th>CLCN6</th>\n",
       "      <th>TNFRSF1B</th>\n",
       "      <th>THAP3</th>\n",
       "      <th>VPS13D</th>\n",
       "      <th>H6PD</th>\n",
       "      <th>VAMP3</th>\n",
       "      <th>...</th>\n",
       "      <th>SHISA8</th>\n",
       "      <th>NFAM1</th>\n",
       "      <th>APOBEC3G</th>\n",
       "      <th>ADSL</th>\n",
       "      <th>PDXP</th>\n",
       "      <th>ARHGAP8</th>\n",
       "      <th>ARFGAP3</th>\n",
       "      <th>APOBEC3C</th>\n",
       "      <th>SHANK3</th>\n",
       "      <th>NOL12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46b4_4d67</th>\n",
       "      <td>5.195173</td>\n",
       "      <td>0.434760</td>\n",
       "      <td>6.682951</td>\n",
       "      <td>-1.647374</td>\n",
       "      <td>4.707818</td>\n",
       "      <td>6.323831</td>\n",
       "      <td>3.853517</td>\n",
       "      <td>6.156486</td>\n",
       "      <td>6.290615</td>\n",
       "      <td>6.600191</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.475513</td>\n",
       "      <td>4.405011</td>\n",
       "      <td>5.296945</td>\n",
       "      <td>2.284840</td>\n",
       "      <td>1.467736</td>\n",
       "      <td>1.062751</td>\n",
       "      <td>6.694004</td>\n",
       "      <td>5.187811</td>\n",
       "      <td>0.978422</td>\n",
       "      <td>2.895713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442a_4ee4</th>\n",
       "      <td>5.009279</td>\n",
       "      <td>0.352068</td>\n",
       "      <td>6.386289</td>\n",
       "      <td>0.444971</td>\n",
       "      <td>5.123592</td>\n",
       "      <td>3.563797</td>\n",
       "      <td>4.333164</td>\n",
       "      <td>6.909673</td>\n",
       "      <td>6.274079</td>\n",
       "      <td>6.806959</td>\n",
       "      <td>...</td>\n",
       "      <td>0.123908</td>\n",
       "      <td>1.732832</td>\n",
       "      <td>5.114328</td>\n",
       "      <td>2.633263</td>\n",
       "      <td>1.265538</td>\n",
       "      <td>1.235494</td>\n",
       "      <td>6.370056</td>\n",
       "      <td>5.418903</td>\n",
       "      <td>1.293322</td>\n",
       "      <td>3.340659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce47_4b11</th>\n",
       "      <td>4.895325</td>\n",
       "      <td>0.863774</td>\n",
       "      <td>6.377870</td>\n",
       "      <td>0.150260</td>\n",
       "      <td>5.271711</td>\n",
       "      <td>3.856383</td>\n",
       "      <td>4.055262</td>\n",
       "      <td>7.052612</td>\n",
       "      <td>6.594367</td>\n",
       "      <td>6.497604</td>\n",
       "      <td>...</td>\n",
       "      <td>0.623454</td>\n",
       "      <td>1.392833</td>\n",
       "      <td>4.348760</td>\n",
       "      <td>2.966214</td>\n",
       "      <td>2.002705</td>\n",
       "      <td>3.333978</td>\n",
       "      <td>6.413285</td>\n",
       "      <td>4.863625</td>\n",
       "      <td>1.412305</td>\n",
       "      <td>3.461423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f868_4cba</th>\n",
       "      <td>5.224931</td>\n",
       "      <td>0.953208</td>\n",
       "      <td>6.497938</td>\n",
       "      <td>-0.837709</td>\n",
       "      <td>4.831947</td>\n",
       "      <td>4.505654</td>\n",
       "      <td>4.119327</td>\n",
       "      <td>6.823517</td>\n",
       "      <td>6.755827</td>\n",
       "      <td>6.842240</td>\n",
       "      <td>...</td>\n",
       "      <td>0.453776</td>\n",
       "      <td>3.445990</td>\n",
       "      <td>4.404184</td>\n",
       "      <td>2.627003</td>\n",
       "      <td>1.731249</td>\n",
       "      <td>3.299381</td>\n",
       "      <td>6.912591</td>\n",
       "      <td>5.122441</td>\n",
       "      <td>0.896869</td>\n",
       "      <td>3.661237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273e_4084</th>\n",
       "      <td>5.204216</td>\n",
       "      <td>1.132661</td>\n",
       "      <td>6.520615</td>\n",
       "      <td>-0.864559</td>\n",
       "      <td>5.548620</td>\n",
       "      <td>6.031218</td>\n",
       "      <td>4.066954</td>\n",
       "      <td>6.607320</td>\n",
       "      <td>6.984108</td>\n",
       "      <td>6.633789</td>\n",
       "      <td>...</td>\n",
       "      <td>1.222403</td>\n",
       "      <td>2.702998</td>\n",
       "      <td>5.541400</td>\n",
       "      <td>2.590201</td>\n",
       "      <td>1.284268</td>\n",
       "      <td>3.111907</td>\n",
       "      <td>6.513805</td>\n",
       "      <td>5.164490</td>\n",
       "      <td>1.114981</td>\n",
       "      <td>3.524019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 13757 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_id       DNAJC11    CDK11A      NADK     MASP2     CLCN6  TNFRSF1B  \\\n",
       "sample_name                                                               \n",
       "46b4_4d67    5.195173  0.434760  6.682951 -1.647374  4.707818  6.323831   \n",
       "442a_4ee4    5.009279  0.352068  6.386289  0.444971  5.123592  3.563797   \n",
       "ce47_4b11    4.895325  0.863774  6.377870  0.150260  5.271711  3.856383   \n",
       "f868_4cba    5.224931  0.953208  6.497938 -0.837709  4.831947  4.505654   \n",
       "273e_4084    5.204216  1.132661  6.520615 -0.864559  5.548620  6.031218   \n",
       "\n",
       "gene_id         THAP3    VPS13D      H6PD     VAMP3  ...    SHISA8     NFAM1  \\\n",
       "sample_name                                          ...                       \n",
       "46b4_4d67    3.853517  6.156486  6.290615  6.600191  ... -0.475513  4.405011   \n",
       "442a_4ee4    4.333164  6.909673  6.274079  6.806959  ...  0.123908  1.732832   \n",
       "ce47_4b11    4.055262  7.052612  6.594367  6.497604  ...  0.623454  1.392833   \n",
       "f868_4cba    4.119327  6.823517  6.755827  6.842240  ...  0.453776  3.445990   \n",
       "273e_4084    4.066954  6.607320  6.984108  6.633789  ...  1.222403  2.702998   \n",
       "\n",
       "gene_id      APOBEC3G      ADSL      PDXP   ARHGAP8   ARFGAP3  APOBEC3C  \\\n",
       "sample_name                                                               \n",
       "46b4_4d67    5.296945  2.284840  1.467736  1.062751  6.694004  5.187811   \n",
       "442a_4ee4    5.114328  2.633263  1.265538  1.235494  6.370056  5.418903   \n",
       "ce47_4b11    4.348760  2.966214  2.002705  3.333978  6.413285  4.863625   \n",
       "f868_4cba    4.404184  2.627003  1.731249  3.299381  6.912591  5.122441   \n",
       "273e_4084    5.541400  2.590201  1.284268  3.111907  6.513805  5.164490   \n",
       "\n",
       "gene_id        SHANK3     NOL12  \n",
       "sample_name                      \n",
       "46b4_4d67    0.978422  2.895713  \n",
       "442a_4ee4    1.293322  3.340659  \n",
       "ce47_4b11    1.412305  3.461423  \n",
       "f868_4cba    0.896869  3.661237  \n",
       "273e_4084    1.114981  3.524019  \n",
       "\n",
       "[5 rows x 13757 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE201955_norm_data.index = GSE201955_norm_data.gene_id\n",
    "GSE201955_norm_data.drop(\"gene_id\", axis = 1, inplace = True)\n",
    "GSE201955_norm_data = GSE201955_norm_data.T\n",
    "GSE201955_norm_data.index.name = \"sample_name\"\n",
    "GSE201955_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_id</th>\n",
       "      <th>1/2-SBSRNA4</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2LD1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>tAKR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S-001167453</th>\n",
       "      <td>0.486713</td>\n",
       "      <td>6.67863</td>\n",
       "      <td>0.800632</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.76746</td>\n",
       "      <td>107.67000</td>\n",
       "      <td>0.021372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71.7415</td>\n",
       "      <td>0.346234</td>\n",
       "      <td>...</td>\n",
       "      <td>1.66841</td>\n",
       "      <td>0.923785</td>\n",
       "      <td>2.48296</td>\n",
       "      <td>3.80731</td>\n",
       "      <td>0.130765</td>\n",
       "      <td>5.32574</td>\n",
       "      <td>182.016</td>\n",
       "      <td>3.33042</td>\n",
       "      <td>9.49594</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167456</th>\n",
       "      <td>0.213607</td>\n",
       "      <td>6.06781</td>\n",
       "      <td>0.976157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.23426</td>\n",
       "      <td>111.04000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.5602</td>\n",
       "      <td>0.016334</td>\n",
       "      <td>...</td>\n",
       "      <td>1.21035</td>\n",
       "      <td>1.213130</td>\n",
       "      <td>2.35309</td>\n",
       "      <td>4.27201</td>\n",
       "      <td>0.040554</td>\n",
       "      <td>5.96152</td>\n",
       "      <td>281.918</td>\n",
       "      <td>4.14323</td>\n",
       "      <td>9.04584</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167457</th>\n",
       "      <td>0.328667</td>\n",
       "      <td>5.66054</td>\n",
       "      <td>1.011490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.97234</td>\n",
       "      <td>447.55900</td>\n",
       "      <td>0.013058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.9927</td>\n",
       "      <td>0.235736</td>\n",
       "      <td>...</td>\n",
       "      <td>2.03526</td>\n",
       "      <td>0.925011</td>\n",
       "      <td>2.02607</td>\n",
       "      <td>4.37955</td>\n",
       "      <td>0.207449</td>\n",
       "      <td>5.09867</td>\n",
       "      <td>323.245</td>\n",
       "      <td>4.52346</td>\n",
       "      <td>8.26247</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167458</th>\n",
       "      <td>0.193905</td>\n",
       "      <td>9.01004</td>\n",
       "      <td>0.831169</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.93006</td>\n",
       "      <td>5.41849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.2220</td>\n",
       "      <td>0.079486</td>\n",
       "      <td>...</td>\n",
       "      <td>2.22260</td>\n",
       "      <td>0.586752</td>\n",
       "      <td>2.10799</td>\n",
       "      <td>4.61071</td>\n",
       "      <td>0.186260</td>\n",
       "      <td>5.97974</td>\n",
       "      <td>346.286</td>\n",
       "      <td>3.83862</td>\n",
       "      <td>8.83471</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167459</th>\n",
       "      <td>0.415307</td>\n",
       "      <td>6.41691</td>\n",
       "      <td>0.848899</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.28378</td>\n",
       "      <td>226.60300</td>\n",
       "      <td>0.011371</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80.5774</td>\n",
       "      <td>0.217933</td>\n",
       "      <td>...</td>\n",
       "      <td>1.74028</td>\n",
       "      <td>0.781153</td>\n",
       "      <td>1.84381</td>\n",
       "      <td>3.97499</td>\n",
       "      <td>0.115428</td>\n",
       "      <td>5.74150</td>\n",
       "      <td>184.021</td>\n",
       "      <td>4.79642</td>\n",
       "      <td>8.24657</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23273 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_id      1/2-SBSRNA4     A1BG  A1BG-AS1  A1CF    A2LD1        A2M  \\\n",
       "sample_name                                                             \n",
       "S-001167453     0.486713  6.67863  0.800632   0.0  2.76746  107.67000   \n",
       "S-001167456     0.213607  6.06781  0.976157   0.0  2.23426  111.04000   \n",
       "S-001167457     0.328667  5.66054  1.011490   0.0  2.97234  447.55900   \n",
       "S-001167458     0.193905  9.01004  0.831169   0.0  2.93006    5.41849   \n",
       "S-001167459     0.415307  6.41691  0.848899   0.0  2.28378  226.60300   \n",
       "\n",
       "gene_id         A2ML1  A2MP1   A4GALT     A4GNT  ...    ZWINT      ZXDA  \\\n",
       "sample_name                                      ...                      \n",
       "S-001167453  0.021372    0.0  71.7415  0.346234  ...  1.66841  0.923785   \n",
       "S-001167456  0.000000    0.0  34.5602  0.016334  ...  1.21035  1.213130   \n",
       "S-001167457  0.013058    0.0  64.9927  0.235736  ...  2.03526  0.925011   \n",
       "S-001167458  0.000000    0.0  30.2220  0.079486  ...  2.22260  0.586752   \n",
       "S-001167459  0.011371    0.0  80.5774  0.217933  ...  1.74028  0.781153   \n",
       "\n",
       "gene_id         ZXDB     ZXDC    ZYG11A   ZYG11B      ZYX    ZZEF1     ZZZ3  \\\n",
       "sample_name                                                                   \n",
       "S-001167453  2.48296  3.80731  0.130765  5.32574  182.016  3.33042  9.49594   \n",
       "S-001167456  2.35309  4.27201  0.040554  5.96152  281.918  4.14323  9.04584   \n",
       "S-001167457  2.02607  4.37955  0.207449  5.09867  323.245  4.52346  8.26247   \n",
       "S-001167458  2.10799  4.61071  0.186260  5.97974  346.286  3.83862  8.83471   \n",
       "S-001167459  1.84381  3.97499  0.115428  5.74150  184.021  4.79642  8.24657   \n",
       "\n",
       "gene_id      tAKR  \n",
       "sample_name        \n",
       "S-001167453   0.0  \n",
       "S-001167456   0.0  \n",
       "S-001167457   0.0  \n",
       "S-001167458   0.0  \n",
       "S-001167459   0.0  \n",
       "\n",
       "[5 rows x 23273 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE58434_norm_data.index = GSE58434_norm_data.gene_id\n",
    "GSE58434_norm_data.drop(\"gene_id\", axis = 1, inplace = True)\n",
    "GSE58434_norm_data = GSE58434_norm_data.T\n",
    "GSE58434_norm_data.index.name = \"sample_name\"\n",
    "GSE58434_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_id</th>\n",
       "      <th>1007_PM_s_at</th>\n",
       "      <th>1053_PM_at</th>\n",
       "      <th>117_PM_at</th>\n",
       "      <th>121_PM_at</th>\n",
       "      <th>1255_PM_g_at</th>\n",
       "      <th>1294_PM_at</th>\n",
       "      <th>1316_PM_at</th>\n",
       "      <th>1320_PM_at</th>\n",
       "      <th>1405_PM_i_at</th>\n",
       "      <th>1431_PM_at</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-TagO-3_at</th>\n",
       "      <th>AFFX-r2-TagO-5_at</th>\n",
       "      <th>AFFX-r2-TagQ-3_at</th>\n",
       "      <th>AFFX-r2-TagQ-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_220</th>\n",
       "      <td>4.58</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.75</td>\n",
       "      <td>5.43</td>\n",
       "      <td>2.63</td>\n",
       "      <td>7.63</td>\n",
       "      <td>4.93</td>\n",
       "      <td>3.61</td>\n",
       "      <td>9.56</td>\n",
       "      <td>3.65</td>\n",
       "      <td>...</td>\n",
       "      <td>2.41</td>\n",
       "      <td>3.90</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.78</td>\n",
       "      <td>7.77</td>\n",
       "      <td>5.45</td>\n",
       "      <td>5.85</td>\n",
       "      <td>2.43</td>\n",
       "      <td>3.05</td>\n",
       "      <td>2.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_251</th>\n",
       "      <td>5.62</td>\n",
       "      <td>6.44</td>\n",
       "      <td>8.44</td>\n",
       "      <td>4.53</td>\n",
       "      <td>2.48</td>\n",
       "      <td>7.44</td>\n",
       "      <td>5.48</td>\n",
       "      <td>3.50</td>\n",
       "      <td>9.33</td>\n",
       "      <td>3.81</td>\n",
       "      <td>...</td>\n",
       "      <td>2.61</td>\n",
       "      <td>3.55</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.62</td>\n",
       "      <td>7.94</td>\n",
       "      <td>5.54</td>\n",
       "      <td>6.01</td>\n",
       "      <td>2.51</td>\n",
       "      <td>3.03</td>\n",
       "      <td>2.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_483</th>\n",
       "      <td>4.90</td>\n",
       "      <td>6.75</td>\n",
       "      <td>8.26</td>\n",
       "      <td>5.04</td>\n",
       "      <td>2.54</td>\n",
       "      <td>7.65</td>\n",
       "      <td>5.18</td>\n",
       "      <td>3.80</td>\n",
       "      <td>10.33</td>\n",
       "      <td>3.97</td>\n",
       "      <td>...</td>\n",
       "      <td>2.64</td>\n",
       "      <td>3.65</td>\n",
       "      <td>2.46</td>\n",
       "      <td>2.77</td>\n",
       "      <td>8.72</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.74</td>\n",
       "      <td>2.52</td>\n",
       "      <td>3.04</td>\n",
       "      <td>2.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_282</th>\n",
       "      <td>4.33</td>\n",
       "      <td>6.57</td>\n",
       "      <td>9.16</td>\n",
       "      <td>5.04</td>\n",
       "      <td>2.71</td>\n",
       "      <td>8.18</td>\n",
       "      <td>5.09</td>\n",
       "      <td>3.62</td>\n",
       "      <td>9.63</td>\n",
       "      <td>3.93</td>\n",
       "      <td>...</td>\n",
       "      <td>2.38</td>\n",
       "      <td>3.59</td>\n",
       "      <td>2.51</td>\n",
       "      <td>2.99</td>\n",
       "      <td>8.40</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.68</td>\n",
       "      <td>2.50</td>\n",
       "      <td>2.96</td>\n",
       "      <td>2.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_021</th>\n",
       "      <td>5.30</td>\n",
       "      <td>6.64</td>\n",
       "      <td>9.29</td>\n",
       "      <td>4.85</td>\n",
       "      <td>2.59</td>\n",
       "      <td>8.05</td>\n",
       "      <td>5.20</td>\n",
       "      <td>3.79</td>\n",
       "      <td>9.69</td>\n",
       "      <td>3.84</td>\n",
       "      <td>...</td>\n",
       "      <td>2.39</td>\n",
       "      <td>3.79</td>\n",
       "      <td>2.99</td>\n",
       "      <td>3.12</td>\n",
       "      <td>9.58</td>\n",
       "      <td>6.89</td>\n",
       "      <td>7.61</td>\n",
       "      <td>2.72</td>\n",
       "      <td>3.13</td>\n",
       "      <td>3.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54715 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_id      1007_PM_s_at  1053_PM_at  117_PM_at  121_PM_at  1255_PM_g_at  \\\n",
       "sample_name                                                                 \n",
       "A_220                4.58        6.72       8.75       5.43          2.63   \n",
       "A_251                5.62        6.44       8.44       4.53          2.48   \n",
       "A_483                4.90        6.75       8.26       5.04          2.54   \n",
       "A_282                4.33        6.57       9.16       5.04          2.71   \n",
       "A_021                5.30        6.64       9.29       4.85          2.59   \n",
       "\n",
       "gene_id      1294_PM_at  1316_PM_at  1320_PM_at  1405_PM_i_at  1431_PM_at  \\\n",
       "sample_name                                                                 \n",
       "A_220              7.63        4.93        3.61          9.56        3.65   \n",
       "A_251              7.44        5.48        3.50          9.33        3.81   \n",
       "A_483              7.65        5.18        3.80         10.33        3.97   \n",
       "A_282              8.18        5.09        3.62          9.63        3.93   \n",
       "A_021              8.05        5.20        3.79          9.69        3.84   \n",
       "\n",
       "gene_id      ...  AFFX-r2-TagO-3_at  AFFX-r2-TagO-5_at  AFFX-r2-TagQ-3_at  \\\n",
       "sample_name  ...                                                            \n",
       "A_220        ...               2.41               3.90               2.52   \n",
       "A_251        ...               2.61               3.55               2.42   \n",
       "A_483        ...               2.64               3.65               2.46   \n",
       "A_282        ...               2.38               3.59               2.51   \n",
       "A_021        ...               2.39               3.79               2.99   \n",
       "\n",
       "gene_id      AFFX-r2-TagQ-5_at  AFFX-ThrX-3_at  AFFX-ThrX-5_at  \\\n",
       "sample_name                                                      \n",
       "A_220                     2.78            7.77            5.45   \n",
       "A_251                     2.62            7.94            5.54   \n",
       "A_483                     2.77            8.72            6.01   \n",
       "A_282                     2.99            8.40            6.01   \n",
       "A_021                     3.12            9.58            6.89   \n",
       "\n",
       "gene_id      AFFX-ThrX-M_at  AFFX-TrpnX-3_at  AFFX-TrpnX-5_at  AFFX-TrpnX-M_at  \n",
       "sample_name                                                                     \n",
       "A_220                  5.85             2.43             3.05             2.65  \n",
       "A_251                  6.01             2.51             3.03             2.50  \n",
       "A_483                  6.74             2.52             3.04             2.77  \n",
       "A_282                  6.68             2.50             2.96             2.72  \n",
       "A_021                  7.61             2.72             3.13             3.03  \n",
       "\n",
       "[5 rows x 54715 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE69683_norm_data.index = GSE69683_norm_data.gene_id\n",
    "GSE69683_norm_data.drop(\"gene_id\", axis = 1, inplace = True)\n",
    "GSE69683_norm_data = GSE69683_norm_data.T\n",
    "GSE69683_norm_data.index.name = \"sample_name\"\n",
    "GSE69683_norm_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate Affimetrix Columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LOC100009676</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>CDKN2BAS</th>\n",
       "      <th>LOC100049716</th>\n",
       "      <th>NAALADL1</th>\n",
       "      <th>ACOT8</th>\n",
       "      <th>ABI1</th>\n",
       "      <th>...</th>\n",
       "      <th>PPP4R1</th>\n",
       "      <th>CDC42</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>ROD1</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>NAT1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 01</th>\n",
       "      <td>5.329072</td>\n",
       "      <td>4.888653</td>\n",
       "      <td>6.717174</td>\n",
       "      <td>4.903333</td>\n",
       "      <td>4.445861</td>\n",
       "      <td>4.808544</td>\n",
       "      <td>6.854104</td>\n",
       "      <td>5.103949</td>\n",
       "      <td>5.927904</td>\n",
       "      <td>9.579241</td>\n",
       "      <td>...</td>\n",
       "      <td>10.117098</td>\n",
       "      <td>7.123103</td>\n",
       "      <td>6.346994</td>\n",
       "      <td>8.523364</td>\n",
       "      <td>5.859315</td>\n",
       "      <td>7.303053</td>\n",
       "      <td>5.222605</td>\n",
       "      <td>9.679513</td>\n",
       "      <td>8.545118</td>\n",
       "      <td>9.604065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 02</th>\n",
       "      <td>5.995749</td>\n",
       "      <td>4.980210</td>\n",
       "      <td>6.641973</td>\n",
       "      <td>4.991354</td>\n",
       "      <td>4.354581</td>\n",
       "      <td>4.639297</td>\n",
       "      <td>6.299639</td>\n",
       "      <td>5.765387</td>\n",
       "      <td>6.345293</td>\n",
       "      <td>8.970563</td>\n",
       "      <td>...</td>\n",
       "      <td>9.711283</td>\n",
       "      <td>7.067920</td>\n",
       "      <td>6.315463</td>\n",
       "      <td>8.517702</td>\n",
       "      <td>5.299672</td>\n",
       "      <td>7.654083</td>\n",
       "      <td>4.410016</td>\n",
       "      <td>8.691703</td>\n",
       "      <td>9.025900</td>\n",
       "      <td>9.388568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 03</th>\n",
       "      <td>5.702682</td>\n",
       "      <td>5.371424</td>\n",
       "      <td>7.234998</td>\n",
       "      <td>5.398429</td>\n",
       "      <td>4.379199</td>\n",
       "      <td>4.752368</td>\n",
       "      <td>6.337287</td>\n",
       "      <td>5.251048</td>\n",
       "      <td>5.843571</td>\n",
       "      <td>9.537878</td>\n",
       "      <td>...</td>\n",
       "      <td>9.499548</td>\n",
       "      <td>7.222755</td>\n",
       "      <td>6.033887</td>\n",
       "      <td>8.550329</td>\n",
       "      <td>5.482077</td>\n",
       "      <td>7.495711</td>\n",
       "      <td>4.537349</td>\n",
       "      <td>7.265700</td>\n",
       "      <td>8.853830</td>\n",
       "      <td>9.350149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, healthy sample 01</th>\n",
       "      <td>5.583717</td>\n",
       "      <td>4.834860</td>\n",
       "      <td>7.108063</td>\n",
       "      <td>5.141136</td>\n",
       "      <td>4.159465</td>\n",
       "      <td>5.291560</td>\n",
       "      <td>6.255015</td>\n",
       "      <td>5.151088</td>\n",
       "      <td>5.987954</td>\n",
       "      <td>9.131043</td>\n",
       "      <td>...</td>\n",
       "      <td>9.964046</td>\n",
       "      <td>7.217349</td>\n",
       "      <td>6.062817</td>\n",
       "      <td>8.621510</td>\n",
       "      <td>5.015522</td>\n",
       "      <td>7.563695</td>\n",
       "      <td>4.613191</td>\n",
       "      <td>9.566222</td>\n",
       "      <td>8.849365</td>\n",
       "      <td>9.172806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 04</th>\n",
       "      <td>5.743467</td>\n",
       "      <td>4.736444</td>\n",
       "      <td>7.148278</td>\n",
       "      <td>4.757029</td>\n",
       "      <td>4.754785</td>\n",
       "      <td>4.487500</td>\n",
       "      <td>6.468263</td>\n",
       "      <td>5.425693</td>\n",
       "      <td>6.053634</td>\n",
       "      <td>9.198382</td>\n",
       "      <td>...</td>\n",
       "      <td>10.091075</td>\n",
       "      <td>6.835439</td>\n",
       "      <td>6.413736</td>\n",
       "      <td>8.540482</td>\n",
       "      <td>5.165816</td>\n",
       "      <td>7.486327</td>\n",
       "      <td>4.374204</td>\n",
       "      <td>9.074473</td>\n",
       "      <td>8.714443</td>\n",
       "      <td>9.346384</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19008 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      LOC100009676      AKT3      MED6  \\\n",
       "sample_name                                                              \n",
       "Airway epithelium, asthma sample 01       5.329072  4.888653  6.717174   \n",
       "Airway epithelium, asthma sample 02       5.995749  4.980210  6.641973   \n",
       "Airway epithelium, asthma sample 03       5.702682  5.371424  7.234998   \n",
       "Airway epithelium, healthy sample 01      5.583717  4.834860  7.108063   \n",
       "Airway epithelium, asthma sample 04       5.743467  4.736444  7.148278   \n",
       "\n",
       "                                         NR2E3   NAALAD2  CDKN2BAS  \\\n",
       "sample_name                                                          \n",
       "Airway epithelium, asthma sample 01   4.903333  4.445861  4.808544   \n",
       "Airway epithelium, asthma sample 02   4.991354  4.354581  4.639297   \n",
       "Airway epithelium, asthma sample 03   5.398429  4.379199  4.752368   \n",
       "Airway epithelium, healthy sample 01  5.141136  4.159465  5.291560   \n",
       "Airway epithelium, asthma sample 04   4.757029  4.754785  4.487500   \n",
       "\n",
       "                                      LOC100049716  NAALADL1     ACOT8  \\\n",
       "sample_name                                                              \n",
       "Airway epithelium, asthma sample 01       6.854104  5.103949  5.927904   \n",
       "Airway epithelium, asthma sample 02       6.299639  5.765387  6.345293   \n",
       "Airway epithelium, asthma sample 03       6.337287  5.251048  5.843571   \n",
       "Airway epithelium, healthy sample 01      6.255015  5.151088  5.987954   \n",
       "Airway epithelium, asthma sample 04       6.468263  5.425693  6.053634   \n",
       "\n",
       "                                          ABI1  ...     PPP4R1     CDC42  \\\n",
       "sample_name                                     ...                        \n",
       "Airway epithelium, asthma sample 01   9.579241  ...  10.117098  7.123103   \n",
       "Airway epithelium, asthma sample 02   8.970563  ...   9.711283  7.067920   \n",
       "Airway epithelium, asthma sample 03   9.537878  ...   9.499548  7.222755   \n",
       "Airway epithelium, healthy sample 01  9.131043  ...   9.964046  7.217349   \n",
       "Airway epithelium, asthma sample 04   9.198382  ...  10.091075  6.835439   \n",
       "\n",
       "                                       SLC12A6      ROD1     KCNE2     DGCR2  \\\n",
       "sample_name                                                                    \n",
       "Airway epithelium, asthma sample 01   6.346994  8.523364  5.859315  7.303053   \n",
       "Airway epithelium, asthma sample 02   6.315463  8.517702  5.299672  7.654083   \n",
       "Airway epithelium, asthma sample 03   6.033887  8.550329  5.482077  7.495711   \n",
       "Airway epithelium, healthy sample 01  6.062817  8.621510  5.015522  7.563695   \n",
       "Airway epithelium, asthma sample 04   6.413736  8.540482  5.165816  7.486327   \n",
       "\n",
       "                                      CASP8AP2      SCO2      CDH1      NAT1  \n",
       "sample_name                                                                   \n",
       "Airway epithelium, asthma sample 01   5.222605  9.679513  8.545118  9.604065  \n",
       "Airway epithelium, asthma sample 02   4.410016  8.691703  9.025900  9.388568  \n",
       "Airway epithelium, asthma sample 03   4.537349  7.265700  8.853830  9.350149  \n",
       "Airway epithelium, healthy sample 01  4.613191  9.566222  8.849365  9.172806  \n",
       "Airway epithelium, asthma sample 04   4.374204  9.074473  8.714443  9.346384  \n",
       "\n",
       "[5 rows x 19008 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate GSE67472_norm_data genes\n",
    "AFF_column_names = GSE67472_norm_data.columns\n",
    "annotated_columns_GSE67472 = list(map(lambda x: Affimetrix_map['ENTREZID_to_ALIAS'][x.split(\"_\")[0]] if x.split(\"_\")[0] in Affimetrix_map['ENTREZID_to_ALIAS'] else x,AFF_column_names))\n",
    "GSE67472_norm_data.columns = annotated_columns_GSE67472\n",
    "GSE67472_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17967"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many genes have been annotated?\n",
    "len(AFF_column_names) - len(set(AFF_column_names).intersection(set(annotated_columns_GSE67472)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DDR1</th>\n",
       "      <th>RFC2</th>\n",
       "      <th>HSPA6</th>\n",
       "      <th>PAX8</th>\n",
       "      <th>GUCA1A</th>\n",
       "      <th>UBA7</th>\n",
       "      <th>THRA</th>\n",
       "      <th>PTPN21</th>\n",
       "      <th>CCL5</th>\n",
       "      <th>CYP2E1</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-TagO-3_at</th>\n",
       "      <th>AFFX-r2-TagO-5_at</th>\n",
       "      <th>AFFX-r2-TagQ-3_at</th>\n",
       "      <th>AFFX-r2-TagQ-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_220</th>\n",
       "      <td>4.58</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.75</td>\n",
       "      <td>5.43</td>\n",
       "      <td>2.63</td>\n",
       "      <td>7.63</td>\n",
       "      <td>4.93</td>\n",
       "      <td>3.61</td>\n",
       "      <td>9.56</td>\n",
       "      <td>3.65</td>\n",
       "      <td>...</td>\n",
       "      <td>2.41</td>\n",
       "      <td>3.90</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.78</td>\n",
       "      <td>7.77</td>\n",
       "      <td>5.45</td>\n",
       "      <td>5.85</td>\n",
       "      <td>2.43</td>\n",
       "      <td>3.05</td>\n",
       "      <td>2.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_251</th>\n",
       "      <td>5.62</td>\n",
       "      <td>6.44</td>\n",
       "      <td>8.44</td>\n",
       "      <td>4.53</td>\n",
       "      <td>2.48</td>\n",
       "      <td>7.44</td>\n",
       "      <td>5.48</td>\n",
       "      <td>3.50</td>\n",
       "      <td>9.33</td>\n",
       "      <td>3.81</td>\n",
       "      <td>...</td>\n",
       "      <td>2.61</td>\n",
       "      <td>3.55</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.62</td>\n",
       "      <td>7.94</td>\n",
       "      <td>5.54</td>\n",
       "      <td>6.01</td>\n",
       "      <td>2.51</td>\n",
       "      <td>3.03</td>\n",
       "      <td>2.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_483</th>\n",
       "      <td>4.90</td>\n",
       "      <td>6.75</td>\n",
       "      <td>8.26</td>\n",
       "      <td>5.04</td>\n",
       "      <td>2.54</td>\n",
       "      <td>7.65</td>\n",
       "      <td>5.18</td>\n",
       "      <td>3.80</td>\n",
       "      <td>10.33</td>\n",
       "      <td>3.97</td>\n",
       "      <td>...</td>\n",
       "      <td>2.64</td>\n",
       "      <td>3.65</td>\n",
       "      <td>2.46</td>\n",
       "      <td>2.77</td>\n",
       "      <td>8.72</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.74</td>\n",
       "      <td>2.52</td>\n",
       "      <td>3.04</td>\n",
       "      <td>2.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_282</th>\n",
       "      <td>4.33</td>\n",
       "      <td>6.57</td>\n",
       "      <td>9.16</td>\n",
       "      <td>5.04</td>\n",
       "      <td>2.71</td>\n",
       "      <td>8.18</td>\n",
       "      <td>5.09</td>\n",
       "      <td>3.62</td>\n",
       "      <td>9.63</td>\n",
       "      <td>3.93</td>\n",
       "      <td>...</td>\n",
       "      <td>2.38</td>\n",
       "      <td>3.59</td>\n",
       "      <td>2.51</td>\n",
       "      <td>2.99</td>\n",
       "      <td>8.40</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.68</td>\n",
       "      <td>2.50</td>\n",
       "      <td>2.96</td>\n",
       "      <td>2.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_021</th>\n",
       "      <td>5.30</td>\n",
       "      <td>6.64</td>\n",
       "      <td>9.29</td>\n",
       "      <td>4.85</td>\n",
       "      <td>2.59</td>\n",
       "      <td>8.05</td>\n",
       "      <td>5.20</td>\n",
       "      <td>3.79</td>\n",
       "      <td>9.69</td>\n",
       "      <td>3.84</td>\n",
       "      <td>...</td>\n",
       "      <td>2.39</td>\n",
       "      <td>3.79</td>\n",
       "      <td>2.99</td>\n",
       "      <td>3.12</td>\n",
       "      <td>9.58</td>\n",
       "      <td>6.89</td>\n",
       "      <td>7.61</td>\n",
       "      <td>2.72</td>\n",
       "      <td>3.13</td>\n",
       "      <td>3.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54715 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             DDR1  RFC2  HSPA6  PAX8  GUCA1A  UBA7  THRA  PTPN21   CCL5  \\\n",
       "sample_name                                                               \n",
       "A_220        4.58  6.72   8.75  5.43    2.63  7.63  4.93    3.61   9.56   \n",
       "A_251        5.62  6.44   8.44  4.53    2.48  7.44  5.48    3.50   9.33   \n",
       "A_483        4.90  6.75   8.26  5.04    2.54  7.65  5.18    3.80  10.33   \n",
       "A_282        4.33  6.57   9.16  5.04    2.71  8.18  5.09    3.62   9.63   \n",
       "A_021        5.30  6.64   9.29  4.85    2.59  8.05  5.20    3.79   9.69   \n",
       "\n",
       "             CYP2E1  ...  AFFX-r2-TagO-3_at  AFFX-r2-TagO-5_at  \\\n",
       "sample_name          ...                                         \n",
       "A_220          3.65  ...               2.41               3.90   \n",
       "A_251          3.81  ...               2.61               3.55   \n",
       "A_483          3.97  ...               2.64               3.65   \n",
       "A_282          3.93  ...               2.38               3.59   \n",
       "A_021          3.84  ...               2.39               3.79   \n",
       "\n",
       "             AFFX-r2-TagQ-3_at  AFFX-r2-TagQ-5_at  AFFX-ThrX-3_at  \\\n",
       "sample_name                                                         \n",
       "A_220                     2.52               2.78            7.77   \n",
       "A_251                     2.42               2.62            7.94   \n",
       "A_483                     2.46               2.77            8.72   \n",
       "A_282                     2.51               2.99            8.40   \n",
       "A_021                     2.99               3.12            9.58   \n",
       "\n",
       "             AFFX-ThrX-5_at  AFFX-ThrX-M_at  AFFX-TrpnX-3_at  AFFX-TrpnX-5_at  \\\n",
       "sample_name                                                                     \n",
       "A_220                  5.45            5.85             2.43             3.05   \n",
       "A_251                  5.54            6.01             2.51             3.03   \n",
       "A_483                  6.01            6.74             2.52             3.04   \n",
       "A_282                  6.01            6.68             2.50             2.96   \n",
       "A_021                  6.89            7.61             2.72             3.13   \n",
       "\n",
       "             AFFX-TrpnX-M_at  \n",
       "sample_name                   \n",
       "A_220                   2.65  \n",
       "A_251                   2.50  \n",
       "A_483                   2.77  \n",
       "A_282                   2.72  \n",
       "A_021                   3.03  \n",
       "\n",
       "[5 rows x 54715 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# annotate GSE69683_norm_data genes\n",
    "AFF_PM_column_names = GSE69683_norm_data.columns\n",
    "annotated_columns_GSE69683 = list(map(lambda x: Affimetrix_map['AFF_PM_to_ALIAS'][x] if x in Affimetrix_map['AFF_PM_to_ALIAS'] else x,AFF_PM_column_names))\n",
    "GSE69683_norm_data.columns = annotated_columns_GSE69683\n",
    "GSE69683_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41791"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many genes have been annotated?\n",
    "len(AFF_PM_column_names) - len(set(AFF_PM_column_names).intersection(set(annotated_columns_GSE69683)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "1q5ZumLVFyjn"
   },
   "source": [
    "### Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "executionInfo": {
     "elapsed": 372,
     "status": "ok",
     "timestamp": 1710581230011,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "1LNYUNqtJZg-"
   },
   "outputs": [],
   "source": [
    "def get_Limma_DEGs(results, fc_thresh=0):\n",
    "    \"\"\"\n",
    "    This function returns a list of all differentially expressed genes\n",
    "    from Limma results. The criteria for selecting genes are:\n",
    "        pvalue < 0.05\n",
    "        abs(logFC) >= fc_thresh\n",
    "    \"\"\"\n",
    "\n",
    "    # Sort results by ascending p-values and descending logFC\n",
    "    sorted_results = results.sort_values(by=[\"P.Value\", \"logFC\"], ascending=[True, False])\n",
    "\n",
    "    # Filter results by logFC and P-value\n",
    "    filtered_results = sorted_results[(sorted_results[\"logFC\"].abs() >= fc_thresh) & (sorted_results[\"P.Value\"] < 0.05)]\n",
    "\n",
    "    # Extract differentially expressed genes\n",
    "    DEGs = filtered_results.iloc[:, 0].values\n",
    "\n",
    "    return DEGs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "executionInfo": {
     "elapsed": 420,
     "status": "ok",
     "timestamp": 1710581233117,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "7EefOvL2Jaf4"
   },
   "outputs": [],
   "source": [
    "def get_coexpressed_genes(module_df, WGCNA_results, r_thresh=0.2, Pvalue_thresh=0.001):\n",
    "    \"\"\"\n",
    "    Function returns a list of co-expressed genes from a given dataset based\n",
    "    on WGCNA results and gene modules correlation with the subject types based\n",
    "    on these criteria:\n",
    "              Pvalue < Pvalue_thresh\n",
    "              |r| >= r_thresh\n",
    "    \"\"\"\n",
    "\n",
    "    # Sort WGCNA results by ascending P-values and descending correlation\n",
    "    sorted_results = WGCNA_results.sort_values(by=[\"Pvalue\", \"Correlation\"], ascending=[True, False])\n",
    "\n",
    "    # Filter WGCNA results\n",
    "    filtered_results = sorted_results[(sorted_results[\"Correlation\"].abs() >= r_thresh) &\n",
    "                                      (sorted_results[\"Pvalue\"] < Pvalue_thresh)]\n",
    "\n",
    "    # Extract unique module names\n",
    "    co_expressed_modules = filtered_results.iloc[:, 0].values\n",
    "\n",
    "    # Filter module_df to only include co-expressed genes\n",
    "    co_expressed_genes = module_df[module_df[\"colors\"].isin([module.split(\"ME\")[-1] for module in co_expressed_modules])][\"gene_id\"].unique()\n",
    "\n",
    "    return list(co_expressed_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "executionInfo": {
     "elapsed": 4,
     "status": "ok",
     "timestamp": 1710581235751,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "hU5-By5sWrPn"
   },
   "outputs": [],
   "source": [
    "def get_differentially_co_expressed_genes(differentially_expressed_genes, co_expressed_genes):\n",
    "    \"\"\"\n",
    "    Function returns genes that are both differentially expressed and co-expressed.\n",
    "    i.e differentially co-expressed.\n",
    "    \"\"\"\n",
    "    differentially_co_expressed_genes = [gene for gene in differentially_expressed_genes if gene in co_expressed_genes]\n",
    "    return differentially_co_expressed_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_column_duplicates(df):\n",
    "    \"\"\"\n",
    "    For one reason or another some dataframes may contain the same column name twice.\n",
    "    This function will automatically select the first occurence of such column.\n",
    "    \"\"\"\n",
    "\n",
    "    selected_iloc_indices = []\n",
    "    seen_columns = []\n",
    "    for i, column in enumerate(df.columns):\n",
    "        if column not in seen_columns:\n",
    "            selected_iloc_indices.append(i)\n",
    "            seen_columns.append(column)\n",
    "\n",
    "    df = df.iloc[:,selected_iloc_indices]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add Labels and Batch to Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HR1005</th>\n",
       "      <td>12.047600</td>\n",
       "      <td>12.047600</td>\n",
       "      <td>0.0</td>\n",
       "      <td>531.815488</td>\n",
       "      <td>75.727772</td>\n",
       "      <td>151.455543</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.047600</td>\n",
       "      <td>7380.015570</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>213.414629</td>\n",
       "      <td>487.067259</td>\n",
       "      <td>1485.296977</td>\n",
       "      <td>15.489771</td>\n",
       "      <td>1519.718691</td>\n",
       "      <td>2039.486579</td>\n",
       "      <td>4042.830358</td>\n",
       "      <td>1089.447261</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1006</th>\n",
       "      <td>16.204622</td>\n",
       "      <td>13.645997</td>\n",
       "      <td>0.0</td>\n",
       "      <td>382.940804</td>\n",
       "      <td>23.880496</td>\n",
       "      <td>633.686008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.705750</td>\n",
       "      <td>20669.421796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>187.632465</td>\n",
       "      <td>508.313406</td>\n",
       "      <td>1739.864678</td>\n",
       "      <td>52.025365</td>\n",
       "      <td>1409.802114</td>\n",
       "      <td>3917.254149</td>\n",
       "      <td>4017.040506</td>\n",
       "      <td>1224.728273</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1009</th>\n",
       "      <td>21.992452</td>\n",
       "      <td>12.936737</td>\n",
       "      <td>0.0</td>\n",
       "      <td>169.471249</td>\n",
       "      <td>29.754494</td>\n",
       "      <td>648.130502</td>\n",
       "      <td>1.293674</td>\n",
       "      <td>9.055716</td>\n",
       "      <td>12044.101737</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>204.400438</td>\n",
       "      <td>512.294768</td>\n",
       "      <td>1503.248788</td>\n",
       "      <td>67.271030</td>\n",
       "      <td>1582.162881</td>\n",
       "      <td>2668.848752</td>\n",
       "      <td>3708.962372</td>\n",
       "      <td>1353.182644</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1010</th>\n",
       "      <td>9.808778</td>\n",
       "      <td>3.772607</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.532202</td>\n",
       "      <td>26.408249</td>\n",
       "      <td>705.477508</td>\n",
       "      <td>0.754521</td>\n",
       "      <td>8.299735</td>\n",
       "      <td>18126.622082</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>202.966256</td>\n",
       "      <td>605.880683</td>\n",
       "      <td>1489.425241</td>\n",
       "      <td>51.307455</td>\n",
       "      <td>1876.494719</td>\n",
       "      <td>3338.757189</td>\n",
       "      <td>3414.963850</td>\n",
       "      <td>1251.751000</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1012</th>\n",
       "      <td>11.376166</td>\n",
       "      <td>21.330311</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1083.579785</td>\n",
       "      <td>28.440414</td>\n",
       "      <td>203.348962</td>\n",
       "      <td>4.266062</td>\n",
       "      <td>5.688083</td>\n",
       "      <td>13180.710004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>214.725128</td>\n",
       "      <td>484.909064</td>\n",
       "      <td>1275.552581</td>\n",
       "      <td>27.018394</td>\n",
       "      <td>1504.497916</td>\n",
       "      <td>3924.777172</td>\n",
       "      <td>3431.335985</td>\n",
       "      <td>1215.827711</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26487 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             A1BG   A1BG-AS1  A1CF          A2M    A2M-AS1       A2ML1  \\\n",
       "HR1005  12.047600  12.047600   0.0   531.815488  75.727772  151.455543   \n",
       "HR1006  16.204622  13.645997   0.0   382.940804  23.880496  633.686008   \n",
       "HR1009  21.992452  12.936737   0.0   169.471249  29.754494  648.130502   \n",
       "HR1010   9.808778   3.772607   0.0   130.532202  26.408249  705.477508   \n",
       "HR1012  11.376166  21.330311   0.0  1083.579785  28.440414  203.348962   \n",
       "\n",
       "           A2MP1    A3GALT2        A4GALT  A4GNT  ...        ZXDA        ZXDB  \\\n",
       "HR1005  0.000000  12.047600   7380.015570    0.0  ...  213.414629  487.067259   \n",
       "HR1006  0.000000   1.705750  20669.421796    0.0  ...  187.632465  508.313406   \n",
       "HR1009  1.293674   9.055716  12044.101737    0.0  ...  204.400438  512.294768   \n",
       "HR1010  0.754521   8.299735  18126.622082    0.0  ...  202.966256  605.880683   \n",
       "HR1012  4.266062   5.688083  13180.710004    0.0  ...  214.725128  484.909064   \n",
       "\n",
       "               ZXDC     ZYG11A       ZYG11B          ZYX        ZZEF1  \\\n",
       "HR1005  1485.296977  15.489771  1519.718691  2039.486579  4042.830358   \n",
       "HR1006  1739.864678  52.025365  1409.802114  3917.254149  4017.040506   \n",
       "HR1009  1503.248788  67.271030  1582.162881  2668.848752  3708.962372   \n",
       "HR1010  1489.425241  51.307455  1876.494719  3338.757189  3414.963850   \n",
       "HR1012  1275.552581  27.018394  1504.497916  3924.777172  3431.335985   \n",
       "\n",
       "               ZZZ3       Type      Batch  \n",
       "HR1005  1089.447261  asthmatic  GSE152004  \n",
       "HR1006  1224.728273  asthmatic  GSE152004  \n",
       "HR1009  1353.182644  asthmatic  GSE152004  \n",
       "HR1010  1251.751000  asthmatic  GSE152004  \n",
       "HR1012  1215.827711  asthmatic  GSE152004  \n",
       "\n",
       "[5 rows x 26487 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE152004_metadata[\"Batch\"] = [\"GSE152004\"]*GSE152004_norm_data.shape[0]\n",
    "GSE152004_norm_data = pd.concat([GSE152004_norm_data,\n",
    "                                 GSE152004_metadata[[\"Type\", \"Batch\"]]\n",
    "                                ], axis =1)\n",
    "GSE152004_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LOC100009676</th>\n",
       "      <th>AKT3</th>\n",
       "      <th>MED6</th>\n",
       "      <th>NR2E3</th>\n",
       "      <th>NAALAD2</th>\n",
       "      <th>CDKN2BAS</th>\n",
       "      <th>LOC100049716</th>\n",
       "      <th>NAALADL1</th>\n",
       "      <th>ACOT8</th>\n",
       "      <th>ABI1</th>\n",
       "      <th>...</th>\n",
       "      <th>SLC12A6</th>\n",
       "      <th>ROD1</th>\n",
       "      <th>KCNE2</th>\n",
       "      <th>DGCR2</th>\n",
       "      <th>CASP8AP2</th>\n",
       "      <th>SCO2</th>\n",
       "      <th>CDH1</th>\n",
       "      <th>NAT1</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 01</th>\n",
       "      <td>5.329072</td>\n",
       "      <td>4.888653</td>\n",
       "      <td>6.717174</td>\n",
       "      <td>4.903333</td>\n",
       "      <td>4.445861</td>\n",
       "      <td>4.808544</td>\n",
       "      <td>6.854104</td>\n",
       "      <td>5.103949</td>\n",
       "      <td>5.927904</td>\n",
       "      <td>9.579241</td>\n",
       "      <td>...</td>\n",
       "      <td>6.346994</td>\n",
       "      <td>8.523364</td>\n",
       "      <td>5.859315</td>\n",
       "      <td>7.303053</td>\n",
       "      <td>5.222605</td>\n",
       "      <td>9.679513</td>\n",
       "      <td>8.545118</td>\n",
       "      <td>9.604065</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE67472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 02</th>\n",
       "      <td>5.995749</td>\n",
       "      <td>4.980210</td>\n",
       "      <td>6.641973</td>\n",
       "      <td>4.991354</td>\n",
       "      <td>4.354581</td>\n",
       "      <td>4.639297</td>\n",
       "      <td>6.299639</td>\n",
       "      <td>5.765387</td>\n",
       "      <td>6.345293</td>\n",
       "      <td>8.970563</td>\n",
       "      <td>...</td>\n",
       "      <td>6.315463</td>\n",
       "      <td>8.517702</td>\n",
       "      <td>5.299672</td>\n",
       "      <td>7.654083</td>\n",
       "      <td>4.410016</td>\n",
       "      <td>8.691703</td>\n",
       "      <td>9.025900</td>\n",
       "      <td>9.388568</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE67472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 03</th>\n",
       "      <td>5.702682</td>\n",
       "      <td>5.371424</td>\n",
       "      <td>7.234998</td>\n",
       "      <td>5.398429</td>\n",
       "      <td>4.379199</td>\n",
       "      <td>4.752368</td>\n",
       "      <td>6.337287</td>\n",
       "      <td>5.251048</td>\n",
       "      <td>5.843571</td>\n",
       "      <td>9.537878</td>\n",
       "      <td>...</td>\n",
       "      <td>6.033887</td>\n",
       "      <td>8.550329</td>\n",
       "      <td>5.482077</td>\n",
       "      <td>7.495711</td>\n",
       "      <td>4.537349</td>\n",
       "      <td>7.265700</td>\n",
       "      <td>8.853830</td>\n",
       "      <td>9.350149</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE67472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, healthy sample 01</th>\n",
       "      <td>5.583717</td>\n",
       "      <td>4.834860</td>\n",
       "      <td>7.108063</td>\n",
       "      <td>5.141136</td>\n",
       "      <td>4.159465</td>\n",
       "      <td>5.291560</td>\n",
       "      <td>6.255015</td>\n",
       "      <td>5.151088</td>\n",
       "      <td>5.987954</td>\n",
       "      <td>9.131043</td>\n",
       "      <td>...</td>\n",
       "      <td>6.062817</td>\n",
       "      <td>8.621510</td>\n",
       "      <td>5.015522</td>\n",
       "      <td>7.563695</td>\n",
       "      <td>4.613191</td>\n",
       "      <td>9.566222</td>\n",
       "      <td>8.849365</td>\n",
       "      <td>9.172806</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE67472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Airway epithelium, asthma sample 04</th>\n",
       "      <td>5.743467</td>\n",
       "      <td>4.736444</td>\n",
       "      <td>7.148278</td>\n",
       "      <td>4.757029</td>\n",
       "      <td>4.754785</td>\n",
       "      <td>4.487500</td>\n",
       "      <td>6.468263</td>\n",
       "      <td>5.425693</td>\n",
       "      <td>6.053634</td>\n",
       "      <td>9.198382</td>\n",
       "      <td>...</td>\n",
       "      <td>6.413736</td>\n",
       "      <td>8.540482</td>\n",
       "      <td>5.165816</td>\n",
       "      <td>7.486327</td>\n",
       "      <td>4.374204</td>\n",
       "      <td>9.074473</td>\n",
       "      <td>8.714443</td>\n",
       "      <td>9.346384</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE67472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19010 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      LOC100009676      AKT3      MED6  \\\n",
       "Airway epithelium, asthma sample 01       5.329072  4.888653  6.717174   \n",
       "Airway epithelium, asthma sample 02       5.995749  4.980210  6.641973   \n",
       "Airway epithelium, asthma sample 03       5.702682  5.371424  7.234998   \n",
       "Airway epithelium, healthy sample 01      5.583717  4.834860  7.108063   \n",
       "Airway epithelium, asthma sample 04       5.743467  4.736444  7.148278   \n",
       "\n",
       "                                         NR2E3   NAALAD2  CDKN2BAS  \\\n",
       "Airway epithelium, asthma sample 01   4.903333  4.445861  4.808544   \n",
       "Airway epithelium, asthma sample 02   4.991354  4.354581  4.639297   \n",
       "Airway epithelium, asthma sample 03   5.398429  4.379199  4.752368   \n",
       "Airway epithelium, healthy sample 01  5.141136  4.159465  5.291560   \n",
       "Airway epithelium, asthma sample 04   4.757029  4.754785  4.487500   \n",
       "\n",
       "                                      LOC100049716  NAALADL1     ACOT8  \\\n",
       "Airway epithelium, asthma sample 01       6.854104  5.103949  5.927904   \n",
       "Airway epithelium, asthma sample 02       6.299639  5.765387  6.345293   \n",
       "Airway epithelium, asthma sample 03       6.337287  5.251048  5.843571   \n",
       "Airway epithelium, healthy sample 01      6.255015  5.151088  5.987954   \n",
       "Airway epithelium, asthma sample 04       6.468263  5.425693  6.053634   \n",
       "\n",
       "                                          ABI1  ...   SLC12A6      ROD1  \\\n",
       "Airway epithelium, asthma sample 01   9.579241  ...  6.346994  8.523364   \n",
       "Airway epithelium, asthma sample 02   8.970563  ...  6.315463  8.517702   \n",
       "Airway epithelium, asthma sample 03   9.537878  ...  6.033887  8.550329   \n",
       "Airway epithelium, healthy sample 01  9.131043  ...  6.062817  8.621510   \n",
       "Airway epithelium, asthma sample 04   9.198382  ...  6.413736  8.540482   \n",
       "\n",
       "                                         KCNE2     DGCR2  CASP8AP2      SCO2  \\\n",
       "Airway epithelium, asthma sample 01   5.859315  7.303053  5.222605  9.679513   \n",
       "Airway epithelium, asthma sample 02   5.299672  7.654083  4.410016  8.691703   \n",
       "Airway epithelium, asthma sample 03   5.482077  7.495711  4.537349  7.265700   \n",
       "Airway epithelium, healthy sample 01  5.015522  7.563695  4.613191  9.566222   \n",
       "Airway epithelium, asthma sample 04   5.165816  7.486327  4.374204  9.074473   \n",
       "\n",
       "                                          CDH1      NAT1             Type  \\\n",
       "Airway epithelium, asthma sample 01   8.545118  9.604065        asthmatic   \n",
       "Airway epithelium, asthma sample 02   9.025900  9.388568        asthmatic   \n",
       "Airway epithelium, asthma sample 03   8.853830  9.350149        asthmatic   \n",
       "Airway epithelium, healthy sample 01  8.849365  9.172806  healthy control   \n",
       "Airway epithelium, asthma sample 04   8.714443  9.346384        asthmatic   \n",
       "\n",
       "                                         Batch  \n",
       "Airway epithelium, asthma sample 01   GSE67472  \n",
       "Airway epithelium, asthma sample 02   GSE67472  \n",
       "Airway epithelium, asthma sample 03   GSE67472  \n",
       "Airway epithelium, healthy sample 01  GSE67472  \n",
       "Airway epithelium, asthma sample 04   GSE67472  \n",
       "\n",
       "[5 rows x 19010 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE67472_metadata[\"Batch\"] = [\"GSE67472\"]*GSE67472_norm_data.shape[0]\n",
    "GSE67472_norm_data = pd.concat([GSE67472_norm_data,\n",
    "                                 GSE67472_metadata[[\"Type\", \"Batch\"]]\n",
    "                                ], axis =1)\n",
    "GSE67472_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DNAJC11</th>\n",
       "      <th>CDK11A</th>\n",
       "      <th>NADK</th>\n",
       "      <th>MASP2</th>\n",
       "      <th>CLCN6</th>\n",
       "      <th>TNFRSF1B</th>\n",
       "      <th>THAP3</th>\n",
       "      <th>VPS13D</th>\n",
       "      <th>H6PD</th>\n",
       "      <th>VAMP3</th>\n",
       "      <th>...</th>\n",
       "      <th>APOBEC3G</th>\n",
       "      <th>ADSL</th>\n",
       "      <th>PDXP</th>\n",
       "      <th>ARHGAP8</th>\n",
       "      <th>ARFGAP3</th>\n",
       "      <th>APOBEC3C</th>\n",
       "      <th>SHANK3</th>\n",
       "      <th>NOL12</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46b4_4d67</th>\n",
       "      <td>5.195173</td>\n",
       "      <td>0.434760</td>\n",
       "      <td>6.682951</td>\n",
       "      <td>-1.647374</td>\n",
       "      <td>4.707818</td>\n",
       "      <td>6.323831</td>\n",
       "      <td>3.853517</td>\n",
       "      <td>6.156486</td>\n",
       "      <td>6.290615</td>\n",
       "      <td>6.600191</td>\n",
       "      <td>...</td>\n",
       "      <td>5.296945</td>\n",
       "      <td>2.284840</td>\n",
       "      <td>1.467736</td>\n",
       "      <td>1.062751</td>\n",
       "      <td>6.694004</td>\n",
       "      <td>5.187811</td>\n",
       "      <td>0.978422</td>\n",
       "      <td>2.895713</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE201955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442a_4ee4</th>\n",
       "      <td>5.009279</td>\n",
       "      <td>0.352068</td>\n",
       "      <td>6.386289</td>\n",
       "      <td>0.444971</td>\n",
       "      <td>5.123592</td>\n",
       "      <td>3.563797</td>\n",
       "      <td>4.333164</td>\n",
       "      <td>6.909673</td>\n",
       "      <td>6.274079</td>\n",
       "      <td>6.806959</td>\n",
       "      <td>...</td>\n",
       "      <td>5.114328</td>\n",
       "      <td>2.633263</td>\n",
       "      <td>1.265538</td>\n",
       "      <td>1.235494</td>\n",
       "      <td>6.370056</td>\n",
       "      <td>5.418903</td>\n",
       "      <td>1.293322</td>\n",
       "      <td>3.340659</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE201955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ce47_4b11</th>\n",
       "      <td>4.895325</td>\n",
       "      <td>0.863774</td>\n",
       "      <td>6.377870</td>\n",
       "      <td>0.150260</td>\n",
       "      <td>5.271711</td>\n",
       "      <td>3.856383</td>\n",
       "      <td>4.055262</td>\n",
       "      <td>7.052612</td>\n",
       "      <td>6.594367</td>\n",
       "      <td>6.497604</td>\n",
       "      <td>...</td>\n",
       "      <td>4.348760</td>\n",
       "      <td>2.966214</td>\n",
       "      <td>2.002705</td>\n",
       "      <td>3.333978</td>\n",
       "      <td>6.413285</td>\n",
       "      <td>4.863625</td>\n",
       "      <td>1.412305</td>\n",
       "      <td>3.461423</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE201955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f868_4cba</th>\n",
       "      <td>5.224931</td>\n",
       "      <td>0.953208</td>\n",
       "      <td>6.497938</td>\n",
       "      <td>-0.837709</td>\n",
       "      <td>4.831947</td>\n",
       "      <td>4.505654</td>\n",
       "      <td>4.119327</td>\n",
       "      <td>6.823517</td>\n",
       "      <td>6.755827</td>\n",
       "      <td>6.842240</td>\n",
       "      <td>...</td>\n",
       "      <td>4.404184</td>\n",
       "      <td>2.627003</td>\n",
       "      <td>1.731249</td>\n",
       "      <td>3.299381</td>\n",
       "      <td>6.912591</td>\n",
       "      <td>5.122441</td>\n",
       "      <td>0.896869</td>\n",
       "      <td>3.661237</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE201955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273e_4084</th>\n",
       "      <td>5.204216</td>\n",
       "      <td>1.132661</td>\n",
       "      <td>6.520615</td>\n",
       "      <td>-0.864559</td>\n",
       "      <td>5.548620</td>\n",
       "      <td>6.031218</td>\n",
       "      <td>4.066954</td>\n",
       "      <td>6.607320</td>\n",
       "      <td>6.984108</td>\n",
       "      <td>6.633789</td>\n",
       "      <td>...</td>\n",
       "      <td>5.541400</td>\n",
       "      <td>2.590201</td>\n",
       "      <td>1.284268</td>\n",
       "      <td>3.111907</td>\n",
       "      <td>6.513805</td>\n",
       "      <td>5.164490</td>\n",
       "      <td>1.114981</td>\n",
       "      <td>3.524019</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE201955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 13759 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            DNAJC11    CDK11A      NADK     MASP2     CLCN6  TNFRSF1B  \\\n",
       "46b4_4d67  5.195173  0.434760  6.682951 -1.647374  4.707818  6.323831   \n",
       "442a_4ee4  5.009279  0.352068  6.386289  0.444971  5.123592  3.563797   \n",
       "ce47_4b11  4.895325  0.863774  6.377870  0.150260  5.271711  3.856383   \n",
       "f868_4cba  5.224931  0.953208  6.497938 -0.837709  4.831947  4.505654   \n",
       "273e_4084  5.204216  1.132661  6.520615 -0.864559  5.548620  6.031218   \n",
       "\n",
       "              THAP3    VPS13D      H6PD     VAMP3  ...  APOBEC3G      ADSL  \\\n",
       "46b4_4d67  3.853517  6.156486  6.290615  6.600191  ...  5.296945  2.284840   \n",
       "442a_4ee4  4.333164  6.909673  6.274079  6.806959  ...  5.114328  2.633263   \n",
       "ce47_4b11  4.055262  7.052612  6.594367  6.497604  ...  4.348760  2.966214   \n",
       "f868_4cba  4.119327  6.823517  6.755827  6.842240  ...  4.404184  2.627003   \n",
       "273e_4084  4.066954  6.607320  6.984108  6.633789  ...  5.541400  2.590201   \n",
       "\n",
       "               PDXP   ARHGAP8   ARFGAP3  APOBEC3C    SHANK3     NOL12  \\\n",
       "46b4_4d67  1.467736  1.062751  6.694004  5.187811  0.978422  2.895713   \n",
       "442a_4ee4  1.265538  1.235494  6.370056  5.418903  1.293322  3.340659   \n",
       "ce47_4b11  2.002705  3.333978  6.413285  4.863625  1.412305  3.461423   \n",
       "f868_4cba  1.731249  3.299381  6.912591  5.122441  0.896869  3.661237   \n",
       "273e_4084  1.284268  3.111907  6.513805  5.164490  1.114981  3.524019   \n",
       "\n",
       "                      Type      Batch  \n",
       "46b4_4d67        asthmatic  GSE201955  \n",
       "442a_4ee4  healthy control  GSE201955  \n",
       "ce47_4b11        asthmatic  GSE201955  \n",
       "f868_4cba  healthy control  GSE201955  \n",
       "273e_4084        asthmatic  GSE201955  \n",
       "\n",
       "[5 rows x 13759 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE201955_metadata[\"Batch\"] = [\"GSE201955\"]*GSE201955_norm_data.shape[0]\n",
    "GSE201955_norm_data = pd.concat([GSE201955_norm_data,\n",
    "                                 GSE201955_metadata[[\"Type\", \"Batch\"]]\n",
    "                                ], axis =1)\n",
    "GSE201955_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1/2-SBSRNA4</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2LD1</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>tAKR</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S-001167453</th>\n",
       "      <td>0.486713</td>\n",
       "      <td>6.67863</td>\n",
       "      <td>0.800632</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.76746</td>\n",
       "      <td>107.67000</td>\n",
       "      <td>0.021372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>71.7415</td>\n",
       "      <td>0.346234</td>\n",
       "      <td>...</td>\n",
       "      <td>2.48296</td>\n",
       "      <td>3.80731</td>\n",
       "      <td>0.130765</td>\n",
       "      <td>5.32574</td>\n",
       "      <td>182.016</td>\n",
       "      <td>3.33042</td>\n",
       "      <td>9.49594</td>\n",
       "      <td>0.0</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE58434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167456</th>\n",
       "      <td>0.213607</td>\n",
       "      <td>6.06781</td>\n",
       "      <td>0.976157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.23426</td>\n",
       "      <td>111.04000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.5602</td>\n",
       "      <td>0.016334</td>\n",
       "      <td>...</td>\n",
       "      <td>2.35309</td>\n",
       "      <td>4.27201</td>\n",
       "      <td>0.040554</td>\n",
       "      <td>5.96152</td>\n",
       "      <td>281.918</td>\n",
       "      <td>4.14323</td>\n",
       "      <td>9.04584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE58434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167457</th>\n",
       "      <td>0.328667</td>\n",
       "      <td>5.66054</td>\n",
       "      <td>1.011490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.97234</td>\n",
       "      <td>447.55900</td>\n",
       "      <td>0.013058</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.9927</td>\n",
       "      <td>0.235736</td>\n",
       "      <td>...</td>\n",
       "      <td>2.02607</td>\n",
       "      <td>4.37955</td>\n",
       "      <td>0.207449</td>\n",
       "      <td>5.09867</td>\n",
       "      <td>323.245</td>\n",
       "      <td>4.52346</td>\n",
       "      <td>8.26247</td>\n",
       "      <td>0.0</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE58434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167458</th>\n",
       "      <td>0.193905</td>\n",
       "      <td>9.01004</td>\n",
       "      <td>0.831169</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.93006</td>\n",
       "      <td>5.41849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.2220</td>\n",
       "      <td>0.079486</td>\n",
       "      <td>...</td>\n",
       "      <td>2.10799</td>\n",
       "      <td>4.61071</td>\n",
       "      <td>0.186260</td>\n",
       "      <td>5.97974</td>\n",
       "      <td>346.286</td>\n",
       "      <td>3.83862</td>\n",
       "      <td>8.83471</td>\n",
       "      <td>0.0</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE58434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S-001167459</th>\n",
       "      <td>0.415307</td>\n",
       "      <td>6.41691</td>\n",
       "      <td>0.848899</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.28378</td>\n",
       "      <td>226.60300</td>\n",
       "      <td>0.011371</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80.5774</td>\n",
       "      <td>0.217933</td>\n",
       "      <td>...</td>\n",
       "      <td>1.84381</td>\n",
       "      <td>3.97499</td>\n",
       "      <td>0.115428</td>\n",
       "      <td>5.74150</td>\n",
       "      <td>184.021</td>\n",
       "      <td>4.79642</td>\n",
       "      <td>8.24657</td>\n",
       "      <td>0.0</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE58434</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23275 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             1/2-SBSRNA4     A1BG  A1BG-AS1  A1CF    A2LD1        A2M  \\\n",
       "S-001167453     0.486713  6.67863  0.800632   0.0  2.76746  107.67000   \n",
       "S-001167456     0.213607  6.06781  0.976157   0.0  2.23426  111.04000   \n",
       "S-001167457     0.328667  5.66054  1.011490   0.0  2.97234  447.55900   \n",
       "S-001167458     0.193905  9.01004  0.831169   0.0  2.93006    5.41849   \n",
       "S-001167459     0.415307  6.41691  0.848899   0.0  2.28378  226.60300   \n",
       "\n",
       "                A2ML1  A2MP1   A4GALT     A4GNT  ...     ZXDB     ZXDC  \\\n",
       "S-001167453  0.021372    0.0  71.7415  0.346234  ...  2.48296  3.80731   \n",
       "S-001167456  0.000000    0.0  34.5602  0.016334  ...  2.35309  4.27201   \n",
       "S-001167457  0.013058    0.0  64.9927  0.235736  ...  2.02607  4.37955   \n",
       "S-001167458  0.000000    0.0  30.2220  0.079486  ...  2.10799  4.61071   \n",
       "S-001167459  0.011371    0.0  80.5774  0.217933  ...  1.84381  3.97499   \n",
       "\n",
       "               ZYG11A   ZYG11B      ZYX    ZZEF1     ZZZ3  tAKR  \\\n",
       "S-001167453  0.130765  5.32574  182.016  3.33042  9.49594   0.0   \n",
       "S-001167456  0.040554  5.96152  281.918  4.14323  9.04584   0.0   \n",
       "S-001167457  0.207449  5.09867  323.245  4.52346  8.26247   0.0   \n",
       "S-001167458  0.186260  5.97974  346.286  3.83862  8.83471   0.0   \n",
       "S-001167459  0.115428  5.74150  184.021  4.79642  8.24657   0.0   \n",
       "\n",
       "                        Type     Batch  \n",
       "S-001167453        asthmatic  GSE58434  \n",
       "S-001167456  healthy control  GSE58434  \n",
       "S-001167457  healthy control  GSE58434  \n",
       "S-001167458        asthmatic  GSE58434  \n",
       "S-001167459        asthmatic  GSE58434  \n",
       "\n",
       "[5 rows x 23275 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE58434_metadata[\"Batch\"] = [\"GSE58434\"]*GSE58434_norm_data.shape[0]\n",
    "GSE58434_norm_data = pd.concat([GSE58434_norm_data,\n",
    "                                 GSE58434_metadata[[\"Type\", \"Batch\"]]\n",
    "                                ], axis =1)\n",
    "GSE58434_norm_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DDR1</th>\n",
       "      <th>RFC2</th>\n",
       "      <th>HSPA6</th>\n",
       "      <th>PAX8</th>\n",
       "      <th>GUCA1A</th>\n",
       "      <th>UBA7</th>\n",
       "      <th>THRA</th>\n",
       "      <th>PTPN21</th>\n",
       "      <th>CCL5</th>\n",
       "      <th>CYP2E1</th>\n",
       "      <th>...</th>\n",
       "      <th>AFFX-r2-TagQ-3_at</th>\n",
       "      <th>AFFX-r2-TagQ-5_at</th>\n",
       "      <th>AFFX-ThrX-3_at</th>\n",
       "      <th>AFFX-ThrX-5_at</th>\n",
       "      <th>AFFX-ThrX-M_at</th>\n",
       "      <th>AFFX-TrpnX-3_at</th>\n",
       "      <th>AFFX-TrpnX-5_at</th>\n",
       "      <th>AFFX-TrpnX-M_at</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_220</th>\n",
       "      <td>4.58</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.75</td>\n",
       "      <td>5.43</td>\n",
       "      <td>2.63</td>\n",
       "      <td>7.63</td>\n",
       "      <td>4.93</td>\n",
       "      <td>3.61</td>\n",
       "      <td>9.56</td>\n",
       "      <td>3.65</td>\n",
       "      <td>...</td>\n",
       "      <td>2.52</td>\n",
       "      <td>2.78</td>\n",
       "      <td>7.77</td>\n",
       "      <td>5.45</td>\n",
       "      <td>5.85</td>\n",
       "      <td>2.43</td>\n",
       "      <td>3.05</td>\n",
       "      <td>2.65</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE69683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_251</th>\n",
       "      <td>5.62</td>\n",
       "      <td>6.44</td>\n",
       "      <td>8.44</td>\n",
       "      <td>4.53</td>\n",
       "      <td>2.48</td>\n",
       "      <td>7.44</td>\n",
       "      <td>5.48</td>\n",
       "      <td>3.50</td>\n",
       "      <td>9.33</td>\n",
       "      <td>3.81</td>\n",
       "      <td>...</td>\n",
       "      <td>2.42</td>\n",
       "      <td>2.62</td>\n",
       "      <td>7.94</td>\n",
       "      <td>5.54</td>\n",
       "      <td>6.01</td>\n",
       "      <td>2.51</td>\n",
       "      <td>3.03</td>\n",
       "      <td>2.50</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE69683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_483</th>\n",
       "      <td>4.90</td>\n",
       "      <td>6.75</td>\n",
       "      <td>8.26</td>\n",
       "      <td>5.04</td>\n",
       "      <td>2.54</td>\n",
       "      <td>7.65</td>\n",
       "      <td>5.18</td>\n",
       "      <td>3.80</td>\n",
       "      <td>10.33</td>\n",
       "      <td>3.97</td>\n",
       "      <td>...</td>\n",
       "      <td>2.46</td>\n",
       "      <td>2.77</td>\n",
       "      <td>8.72</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.74</td>\n",
       "      <td>2.52</td>\n",
       "      <td>3.04</td>\n",
       "      <td>2.77</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE69683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_282</th>\n",
       "      <td>4.33</td>\n",
       "      <td>6.57</td>\n",
       "      <td>9.16</td>\n",
       "      <td>5.04</td>\n",
       "      <td>2.71</td>\n",
       "      <td>8.18</td>\n",
       "      <td>5.09</td>\n",
       "      <td>3.62</td>\n",
       "      <td>9.63</td>\n",
       "      <td>3.93</td>\n",
       "      <td>...</td>\n",
       "      <td>2.51</td>\n",
       "      <td>2.99</td>\n",
       "      <td>8.40</td>\n",
       "      <td>6.01</td>\n",
       "      <td>6.68</td>\n",
       "      <td>2.50</td>\n",
       "      <td>2.96</td>\n",
       "      <td>2.72</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE69683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_021</th>\n",
       "      <td>5.30</td>\n",
       "      <td>6.64</td>\n",
       "      <td>9.29</td>\n",
       "      <td>4.85</td>\n",
       "      <td>2.59</td>\n",
       "      <td>8.05</td>\n",
       "      <td>5.20</td>\n",
       "      <td>3.79</td>\n",
       "      <td>9.69</td>\n",
       "      <td>3.84</td>\n",
       "      <td>...</td>\n",
       "      <td>2.99</td>\n",
       "      <td>3.12</td>\n",
       "      <td>9.58</td>\n",
       "      <td>6.89</td>\n",
       "      <td>7.61</td>\n",
       "      <td>2.72</td>\n",
       "      <td>3.13</td>\n",
       "      <td>3.03</td>\n",
       "      <td>healthy control</td>\n",
       "      <td>GSE69683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54717 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DDR1  RFC2  HSPA6  PAX8  GUCA1A  UBA7  THRA  PTPN21   CCL5  CYP2E1  \\\n",
       "A_220  4.58  6.72   8.75  5.43    2.63  7.63  4.93    3.61   9.56    3.65   \n",
       "A_251  5.62  6.44   8.44  4.53    2.48  7.44  5.48    3.50   9.33    3.81   \n",
       "A_483  4.90  6.75   8.26  5.04    2.54  7.65  5.18    3.80  10.33    3.97   \n",
       "A_282  4.33  6.57   9.16  5.04    2.71  8.18  5.09    3.62   9.63    3.93   \n",
       "A_021  5.30  6.64   9.29  4.85    2.59  8.05  5.20    3.79   9.69    3.84   \n",
       "\n",
       "       ...  AFFX-r2-TagQ-3_at  AFFX-r2-TagQ-5_at  AFFX-ThrX-3_at  \\\n",
       "A_220  ...               2.52               2.78            7.77   \n",
       "A_251  ...               2.42               2.62            7.94   \n",
       "A_483  ...               2.46               2.77            8.72   \n",
       "A_282  ...               2.51               2.99            8.40   \n",
       "A_021  ...               2.99               3.12            9.58   \n",
       "\n",
       "       AFFX-ThrX-5_at  AFFX-ThrX-M_at  AFFX-TrpnX-3_at  AFFX-TrpnX-5_at  \\\n",
       "A_220            5.45            5.85             2.43             3.05   \n",
       "A_251            5.54            6.01             2.51             3.03   \n",
       "A_483            6.01            6.74             2.52             3.04   \n",
       "A_282            6.01            6.68             2.50             2.96   \n",
       "A_021            6.89            7.61             2.72             3.13   \n",
       "\n",
       "       AFFX-TrpnX-M_at             Type     Batch  \n",
       "A_220             2.65        asthmatic  GSE69683  \n",
       "A_251             2.50        asthmatic  GSE69683  \n",
       "A_483             2.77  healthy control  GSE69683  \n",
       "A_282             2.72        asthmatic  GSE69683  \n",
       "A_021             3.03  healthy control  GSE69683  \n",
       "\n",
       "[5 rows x 54717 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE69683_metadata[\"Batch\"] = [\"GSE69683\"]*GSE69683_norm_data.shape[0]\n",
    "GSE69683_norm_data = pd.concat([GSE69683_norm_data,\n",
    "                                 GSE69683_metadata[[\"Type\", \"Batch\"]]\n",
    "                                ], axis =1)\n",
    "GSE69683_norm_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prepare Data For Experimentation\n",
    "The following datasets are created:  \n",
    "* Including all common genes\n",
    "* Including all common genes that are differentially expressed in either one of the datasets\n",
    "* Including all common genes that are differentially expressed in all the datasets\n",
    "* Including all common genes that are differentially co-expressed in either one of the datasets\n",
    "* Including all common genes that are differentially co-expressed in all the datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 310,
     "status": "ok",
     "timestamp": 1710581215747,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "QYuAW3l8KQop",
    "outputId": "f3cd5851-6767-41ba-f584-31d0490bfffe"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intersection using intersection() method: ['SFTPB', 'SULF1', 'ATP2C1', 'VWA3B', 'PRRX2', 'PPM1A', 'KIDINS220', 'TCF19', 'TP53I11', 'YY1AP1']\n",
      "11544 genes are common in all 5 datasets\n"
     ]
    }
   ],
   "source": [
    "# compute common genes\n",
    "set1 = set(GSE152004_norm_data.columns.unique())\n",
    "set2 = set(GSE201955_norm_data.columns.unique())\n",
    "set3 = set(GSE58434_norm_data.columns.unique())\n",
    "set4 = set(GSE67472_norm_data.columns.unique())\n",
    "set5 = set(GSE69683_norm_data.columns.unique())\n",
    "\n",
    "# Find the intersection using the intersection() method\n",
    "common_genes = list(set1.intersection(set2, set3, set4, set5))\n",
    "print(\"Intersection using intersection() method:\", common_genes[:10])\n",
    "print(f\"{len(common_genes)} genes are common in all 5 datasets\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combine Datasets by common genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SFTPB</th>\n",
       "      <th>SULF1</th>\n",
       "      <th>ATP2C1</th>\n",
       "      <th>VWA3B</th>\n",
       "      <th>PRRX2</th>\n",
       "      <th>PPM1A</th>\n",
       "      <th>KIDINS220</th>\n",
       "      <th>TCF19</th>\n",
       "      <th>TP53I11</th>\n",
       "      <th>YY1AP1</th>\n",
       "      <th>...</th>\n",
       "      <th>RUNX1T1</th>\n",
       "      <th>CUBN</th>\n",
       "      <th>CD24</th>\n",
       "      <th>DCAF4</th>\n",
       "      <th>RGL1</th>\n",
       "      <th>ZNF805</th>\n",
       "      <th>ANKRD10</th>\n",
       "      <th>GABRB2</th>\n",
       "      <th>CDK18</th>\n",
       "      <th>FAM193B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HR1005</th>\n",
       "      <td>46.469314</td>\n",
       "      <td>888.080232</td>\n",
       "      <td>4810.434589</td>\n",
       "      <td>10596.724782</td>\n",
       "      <td>80.891029</td>\n",
       "      <td>2481.805609</td>\n",
       "      <td>3332.021955</td>\n",
       "      <td>316.679773</td>\n",
       "      <td>4808.713504</td>\n",
       "      <td>2168.568008</td>\n",
       "      <td>...</td>\n",
       "      <td>10.326514</td>\n",
       "      <td>32.700629</td>\n",
       "      <td>36724.527108</td>\n",
       "      <td>163.503143</td>\n",
       "      <td>1877.704521</td>\n",
       "      <td>535.257659</td>\n",
       "      <td>784.815089</td>\n",
       "      <td>48.190400</td>\n",
       "      <td>963.808004</td>\n",
       "      <td>444.040116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1006</th>\n",
       "      <td>23.027621</td>\n",
       "      <td>573.131894</td>\n",
       "      <td>6846.879232</td>\n",
       "      <td>4527.912536</td>\n",
       "      <td>166.310594</td>\n",
       "      <td>3084.848304</td>\n",
       "      <td>3590.603085</td>\n",
       "      <td>305.329193</td>\n",
       "      <td>4505.737790</td>\n",
       "      <td>2490.394539</td>\n",
       "      <td>...</td>\n",
       "      <td>7.675874</td>\n",
       "      <td>23.027621</td>\n",
       "      <td>19742.346843</td>\n",
       "      <td>318.122316</td>\n",
       "      <td>1639.225446</td>\n",
       "      <td>544.134149</td>\n",
       "      <td>1322.808880</td>\n",
       "      <td>22.174746</td>\n",
       "      <td>2227.709087</td>\n",
       "      <td>665.242377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1009</th>\n",
       "      <td>59.508988</td>\n",
       "      <td>681.766017</td>\n",
       "      <td>6441.201133</td>\n",
       "      <td>3357.083137</td>\n",
       "      <td>157.828186</td>\n",
       "      <td>2855.137759</td>\n",
       "      <td>3456.696009</td>\n",
       "      <td>257.441058</td>\n",
       "      <td>3360.964158</td>\n",
       "      <td>1919.811705</td>\n",
       "      <td>...</td>\n",
       "      <td>25.873473</td>\n",
       "      <td>75.033072</td>\n",
       "      <td>27198.194942</td>\n",
       "      <td>337.648824</td>\n",
       "      <td>1758.102498</td>\n",
       "      <td>631.312744</td>\n",
       "      <td>1613.211049</td>\n",
       "      <td>33.635515</td>\n",
       "      <td>1457.970210</td>\n",
       "      <td>1106.090976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1010</th>\n",
       "      <td>18.863035</td>\n",
       "      <td>1333.993833</td>\n",
       "      <td>6344.770441</td>\n",
       "      <td>2510.292693</td>\n",
       "      <td>257.291797</td>\n",
       "      <td>2961.496490</td>\n",
       "      <td>3425.527150</td>\n",
       "      <td>261.818925</td>\n",
       "      <td>3759.780130</td>\n",
       "      <td>2209.993177</td>\n",
       "      <td>...</td>\n",
       "      <td>12.072342</td>\n",
       "      <td>112.423688</td>\n",
       "      <td>33152.161215</td>\n",
       "      <td>235.410676</td>\n",
       "      <td>1177.807903</td>\n",
       "      <td>632.288932</td>\n",
       "      <td>2148.876943</td>\n",
       "      <td>27.162770</td>\n",
       "      <td>1117.446191</td>\n",
       "      <td>1175.544339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1012</th>\n",
       "      <td>100.963471</td>\n",
       "      <td>766.469165</td>\n",
       "      <td>4844.824575</td>\n",
       "      <td>2462.939878</td>\n",
       "      <td>263.073832</td>\n",
       "      <td>2710.371482</td>\n",
       "      <td>2727.435731</td>\n",
       "      <td>437.982380</td>\n",
       "      <td>5279.962914</td>\n",
       "      <td>2251.058791</td>\n",
       "      <td>...</td>\n",
       "      <td>9.954145</td>\n",
       "      <td>19.908290</td>\n",
       "      <td>21775.403204</td>\n",
       "      <td>210.459066</td>\n",
       "      <td>2115.966823</td>\n",
       "      <td>507.661395</td>\n",
       "      <td>925.735485</td>\n",
       "      <td>27.018394</td>\n",
       "      <td>893.029009</td>\n",
       "      <td>723.808544</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11544 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             SFTPB        SULF1       ATP2C1         VWA3B       PRRX2  \\\n",
       "HR1005   46.469314   888.080232  4810.434589  10596.724782   80.891029   \n",
       "HR1006   23.027621   573.131894  6846.879232   4527.912536  166.310594   \n",
       "HR1009   59.508988   681.766017  6441.201133   3357.083137  157.828186   \n",
       "HR1010   18.863035  1333.993833  6344.770441   2510.292693  257.291797   \n",
       "HR1012  100.963471   766.469165  4844.824575   2462.939878  263.073832   \n",
       "\n",
       "              PPM1A    KIDINS220       TCF19      TP53I11       YY1AP1  ...  \\\n",
       "HR1005  2481.805609  3332.021955  316.679773  4808.713504  2168.568008  ...   \n",
       "HR1006  3084.848304  3590.603085  305.329193  4505.737790  2490.394539  ...   \n",
       "HR1009  2855.137759  3456.696009  257.441058  3360.964158  1919.811705  ...   \n",
       "HR1010  2961.496490  3425.527150  261.818925  3759.780130  2209.993177  ...   \n",
       "HR1012  2710.371482  2727.435731  437.982380  5279.962914  2251.058791  ...   \n",
       "\n",
       "          RUNX1T1        CUBN          CD24       DCAF4         RGL1  \\\n",
       "HR1005  10.326514   32.700629  36724.527108  163.503143  1877.704521   \n",
       "HR1006   7.675874   23.027621  19742.346843  318.122316  1639.225446   \n",
       "HR1009  25.873473   75.033072  27198.194942  337.648824  1758.102498   \n",
       "HR1010  12.072342  112.423688  33152.161215  235.410676  1177.807903   \n",
       "HR1012   9.954145   19.908290  21775.403204  210.459066  2115.966823   \n",
       "\n",
       "            ZNF805      ANKRD10     GABRB2        CDK18      FAM193B  \n",
       "HR1005  535.257659   784.815089  48.190400   963.808004   444.040116  \n",
       "HR1006  544.134149  1322.808880  22.174746  2227.709087   665.242377  \n",
       "HR1009  631.312744  1613.211049  33.635515  1457.970210  1106.090976  \n",
       "HR1010  632.288932  2148.876943  27.162770  1117.446191  1175.544339  \n",
       "HR1012  507.661395   925.735485  27.018394   893.029009   723.808544  \n",
       "\n",
       "[5 rows x 11544 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all common genes dataset\n",
    "all_common_genes_data = pd.concat([\n",
    "    remove_column_duplicates(GSE152004_norm_data[common_genes+[\"Type\", \"Batch\"]]),\n",
    "    remove_column_duplicates(GSE201955_norm_data[common_genes+[\"Type\", \"Batch\"]]),\n",
    "    remove_column_duplicates(GSE58434_norm_data[common_genes+[\"Type\", \"Batch\"]]),\n",
    "    remove_column_duplicates(GSE67472_norm_data[common_genes+[\"Type\", \"Batch\"]]),\n",
    "    remove_column_duplicates(GSE69683_norm_data[common_genes+[\"Type\", \"Batch\"]])\n",
    "], axis = 0)\n",
    "all_common_genes_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1393, 11544)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_common_genes_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get DEGs for each dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>logFC</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5055_at</td>\n",
       "      <td>2.191252</td>\n",
       "      <td>8.258394</td>\n",
       "      <td>9.070102</td>\n",
       "      <td>5.951979e-15</td>\n",
       "      <td>5.375233e-11</td>\n",
       "      <td>23.424373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1359_at</td>\n",
       "      <td>2.139189</td>\n",
       "      <td>7.452254</td>\n",
       "      <td>8.721566</td>\n",
       "      <td>3.645440e-14</td>\n",
       "      <td>1.646099e-10</td>\n",
       "      <td>21.691637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>80206_at</td>\n",
       "      <td>-1.025014</td>\n",
       "      <td>6.507646</td>\n",
       "      <td>-8.001748</td>\n",
       "      <td>1.481909e-12</td>\n",
       "      <td>2.680203e-09</td>\n",
       "      <td>18.147705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>60437_at</td>\n",
       "      <td>0.567571</td>\n",
       "      <td>7.467006</td>\n",
       "      <td>7.995104</td>\n",
       "      <td>1.532998e-12</td>\n",
       "      <td>2.680203e-09</td>\n",
       "      <td>18.115280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10631_at</td>\n",
       "      <td>2.150195</td>\n",
       "      <td>8.158812</td>\n",
       "      <td>7.959027</td>\n",
       "      <td>1.842553e-12</td>\n",
       "      <td>2.680203e-09</td>\n",
       "      <td>17.939329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0     logFC   AveExpr         t       P.Value     adj.P.Val  \\\n",
       "0    5055_at  2.191252  8.258394  9.070102  5.951979e-15  5.375233e-11   \n",
       "1    1359_at  2.139189  7.452254  8.721566  3.645440e-14  1.646099e-10   \n",
       "2   80206_at -1.025014  6.507646 -8.001748  1.481909e-12  2.680203e-09   \n",
       "3   60437_at  0.567571  7.467006  7.995104  1.532998e-12  2.680203e-09   \n",
       "4   10631_at  2.150195  8.158812  7.959027  1.842553e-12  2.680203e-09   \n",
       "\n",
       "           B  \n",
       "0  23.424373  \n",
       "1  21.691637  \n",
       "2  18.147705  \n",
       "3  18.115280  \n",
       "4  17.939329  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE67472_Limma_Results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>logFC</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SERPINB2</td>\n",
       "      <td>2.191252</td>\n",
       "      <td>8.258394</td>\n",
       "      <td>9.070102</td>\n",
       "      <td>5.951979e-15</td>\n",
       "      <td>5.375233e-11</td>\n",
       "      <td>23.424373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CPA3</td>\n",
       "      <td>2.139189</td>\n",
       "      <td>7.452254</td>\n",
       "      <td>8.721566</td>\n",
       "      <td>3.645440e-14</td>\n",
       "      <td>1.646099e-10</td>\n",
       "      <td>21.691637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FHOD3</td>\n",
       "      <td>-1.025014</td>\n",
       "      <td>6.507646</td>\n",
       "      <td>-8.001748</td>\n",
       "      <td>1.481909e-12</td>\n",
       "      <td>2.680203e-09</td>\n",
       "      <td>18.147705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDH26</td>\n",
       "      <td>0.567571</td>\n",
       "      <td>7.467006</td>\n",
       "      <td>7.995104</td>\n",
       "      <td>1.532998e-12</td>\n",
       "      <td>2.680203e-09</td>\n",
       "      <td>18.115280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>POSTN</td>\n",
       "      <td>2.150195</td>\n",
       "      <td>8.158812</td>\n",
       "      <td>7.959027</td>\n",
       "      <td>1.842553e-12</td>\n",
       "      <td>2.680203e-09</td>\n",
       "      <td>17.939329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0     logFC   AveExpr         t       P.Value     adj.P.Val  \\\n",
       "0   SERPINB2  2.191252  8.258394  9.070102  5.951979e-15  5.375233e-11   \n",
       "1       CPA3  2.139189  7.452254  8.721566  3.645440e-14  1.646099e-10   \n",
       "2      FHOD3 -1.025014  6.507646 -8.001748  1.481909e-12  2.680203e-09   \n",
       "3      CDH26  0.567571  7.467006  7.995104  1.532998e-12  2.680203e-09   \n",
       "4      POSTN  2.150195  8.158812  7.959027  1.842553e-12  2.680203e-09   \n",
       "\n",
       "           B  \n",
       "0  23.424373  \n",
       "1  21.691637  \n",
       "2  18.147705  \n",
       "3  18.115280  \n",
       "4  17.939329  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## First annoatate Affimetrix genes\n",
    "GSE67472_Limma_Results_df[\"Unnamed: 0\"] = GSE67472_Limma_Results_df[\"Unnamed: 0\"].apply(lambda x: Affimetrix_map['ENTREZID_to_ALIAS'][x.split(\"_\")[0]] if x.split(\"_\")[0] in Affimetrix_map['ENTREZID_to_ALIAS'] else x)\n",
    "GSE67472_Limma_Results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>logFC</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241320_PM_at</td>\n",
       "      <td>-0.449565</td>\n",
       "      <td>7.190582</td>\n",
       "      <td>-7.992337</td>\n",
       "      <td>9.234245e-15</td>\n",
       "      <td>4.610566e-10</td>\n",
       "      <td>22.768870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACSL1</td>\n",
       "      <td>0.429497</td>\n",
       "      <td>10.947108</td>\n",
       "      <td>7.506717</td>\n",
       "      <td>2.804013e-13</td>\n",
       "      <td>7.000079e-09</td>\n",
       "      <td>19.525219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>239003_PM_at</td>\n",
       "      <td>-0.350505</td>\n",
       "      <td>7.766245</td>\n",
       "      <td>-7.371571</td>\n",
       "      <td>7.042273e-13</td>\n",
       "      <td>8.355415e-09</td>\n",
       "      <td>18.650976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GPM6B</td>\n",
       "      <td>-0.483698</td>\n",
       "      <td>5.710803</td>\n",
       "      <td>-7.364721</td>\n",
       "      <td>7.376242e-13</td>\n",
       "      <td>8.355415e-09</td>\n",
       "      <td>18.607000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F5</td>\n",
       "      <td>0.572393</td>\n",
       "      <td>6.511707</td>\n",
       "      <td>7.346058</td>\n",
       "      <td>8.367296e-13</td>\n",
       "      <td>8.355415e-09</td>\n",
       "      <td>18.487354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0     logFC    AveExpr         t       P.Value     adj.P.Val  \\\n",
       "0  241320_PM_at -0.449565   7.190582 -7.992337  9.234245e-15  4.610566e-10   \n",
       "1         ACSL1  0.429497  10.947108  7.506717  2.804013e-13  7.000079e-09   \n",
       "2  239003_PM_at -0.350505   7.766245 -7.371571  7.042273e-13  8.355415e-09   \n",
       "3         GPM6B -0.483698   5.710803 -7.364721  7.376242e-13  8.355415e-09   \n",
       "4            F5  0.572393   6.511707  7.346058  8.367296e-13  8.355415e-09   \n",
       "\n",
       "           B  \n",
       "0  22.768870  \n",
       "1  19.525219  \n",
       "2  18.650976  \n",
       "3  18.607000  \n",
       "4  18.487354  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE69683_Limma_Results_df[\"Unnamed: 0\"] = GSE69683_Limma_Results_df[\"Unnamed: 0\"].apply(lambda x: Affimetrix_map['AFF_PM_to_ALIAS'][x] if x in Affimetrix_map['AFF_PM_to_ALIAS'] else x)\n",
    "GSE69683_Limma_Results_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14 385 66 120 86\n"
     ]
    }
   ],
   "source": [
    "# get differentially expressed genes from each dataset assuming a logfoldchage of 0.5\n",
    "deg1 = set(get_Limma_DEGs(GSE152004_Limma_Results_df, fc_thresh=0.5))\n",
    "deg2 = set(get_Limma_DEGs(GSE201955_Limma_Results_df, fc_thresh=0.5))\n",
    "deg3 = set(get_Limma_DEGs(GSE58434_Limma_Results_df, fc_thresh=0.5))\n",
    "deg4 = set(get_Limma_DEGs(GSE67472_Limma_Results_df, fc_thresh=0.5))\n",
    "deg5 = set(get_Limma_DEGs(GSE69683_Limma_Results_df, fc_thresh=0.5))\n",
    "\n",
    "print(len(deg1), len(deg2), len(deg3), len(deg4), len(deg5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All common DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "# get common differentially expressed genes\n",
    "all_common_DEGs = deg1.intersection(deg2,deg3,deg4,deg5)\n",
    "print(len(all_common_DEGs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "455\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['CORO6',\n",
       " 'SULF1',\n",
       " 'LRRC37A3',\n",
       " 'CD22',\n",
       " 'HLA-B',\n",
       " 'CSRNP3',\n",
       " 'EFHD1',\n",
       " 'CYB5R2',\n",
       " 'SLC2A6',\n",
       " 'SCG2']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all differentially expressed genes that have matching annotation\n",
    "all_DEGs = list(deg1.union(deg2,deg3,deg4,deg5).intersection(common_genes))\n",
    "print(len(all_DEGs))\n",
    "all_DEGs[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CORO6</th>\n",
       "      <th>SULF1</th>\n",
       "      <th>LRRC37A3</th>\n",
       "      <th>CD22</th>\n",
       "      <th>HLA-B</th>\n",
       "      <th>CSRNP3</th>\n",
       "      <th>EFHD1</th>\n",
       "      <th>CYB5R2</th>\n",
       "      <th>SLC2A6</th>\n",
       "      <th>SCG2</th>\n",
       "      <th>...</th>\n",
       "      <th>CADPS</th>\n",
       "      <th>RAC2</th>\n",
       "      <th>HSF4</th>\n",
       "      <th>CD24</th>\n",
       "      <th>PGLYRP4</th>\n",
       "      <th>NTS</th>\n",
       "      <th>HBB</th>\n",
       "      <th>PHLDB2</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HR1005</th>\n",
       "      <td>1.721086</td>\n",
       "      <td>888.080232</td>\n",
       "      <td>247.836344</td>\n",
       "      <td>34.421714</td>\n",
       "      <td>50704.906417</td>\n",
       "      <td>123.918172</td>\n",
       "      <td>211.693544</td>\n",
       "      <td>616.148688</td>\n",
       "      <td>25.816286</td>\n",
       "      <td>127.360343</td>\n",
       "      <td>...</td>\n",
       "      <td>8.605429</td>\n",
       "      <td>304.632173</td>\n",
       "      <td>67.122343</td>\n",
       "      <td>36724.527108</td>\n",
       "      <td>30.979543</td>\n",
       "      <td>9047.747634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>166.945315</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1006</th>\n",
       "      <td>4.264374</td>\n",
       "      <td>573.131894</td>\n",
       "      <td>147.547348</td>\n",
       "      <td>133.901350</td>\n",
       "      <td>63955.378653</td>\n",
       "      <td>12.793123</td>\n",
       "      <td>193.602589</td>\n",
       "      <td>1148.822412</td>\n",
       "      <td>51.172491</td>\n",
       "      <td>19.616121</td>\n",
       "      <td>...</td>\n",
       "      <td>14.498872</td>\n",
       "      <td>814.495474</td>\n",
       "      <td>150.105972</td>\n",
       "      <td>19742.346843</td>\n",
       "      <td>23.027621</td>\n",
       "      <td>4201.261472</td>\n",
       "      <td>6.822999</td>\n",
       "      <td>894.665709</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1009</th>\n",
       "      <td>6.468368</td>\n",
       "      <td>681.766017</td>\n",
       "      <td>238.035953</td>\n",
       "      <td>227.686563</td>\n",
       "      <td>56408.052419</td>\n",
       "      <td>87.969809</td>\n",
       "      <td>190.170027</td>\n",
       "      <td>922.389317</td>\n",
       "      <td>60.802662</td>\n",
       "      <td>42.691231</td>\n",
       "      <td>...</td>\n",
       "      <td>31.048168</td>\n",
       "      <td>459.254148</td>\n",
       "      <td>326.005761</td>\n",
       "      <td>27198.194942</td>\n",
       "      <td>9.055716</td>\n",
       "      <td>4655.931488</td>\n",
       "      <td>7.762042</td>\n",
       "      <td>724.457247</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1010</th>\n",
       "      <td>14.335907</td>\n",
       "      <td>1333.993833</td>\n",
       "      <td>219.565727</td>\n",
       "      <td>43.762241</td>\n",
       "      <td>40282.388433</td>\n",
       "      <td>30.935377</td>\n",
       "      <td>418.004855</td>\n",
       "      <td>768.857305</td>\n",
       "      <td>38.480591</td>\n",
       "      <td>36.971549</td>\n",
       "      <td>...</td>\n",
       "      <td>27.917292</td>\n",
       "      <td>368.206443</td>\n",
       "      <td>390.087563</td>\n",
       "      <td>33152.161215</td>\n",
       "      <td>11.317821</td>\n",
       "      <td>3876.730946</td>\n",
       "      <td>14.335907</td>\n",
       "      <td>1795.006407</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1012</th>\n",
       "      <td>5.688083</td>\n",
       "      <td>766.469165</td>\n",
       "      <td>191.972796</td>\n",
       "      <td>69.679015</td>\n",
       "      <td>207430.161649</td>\n",
       "      <td>45.504663</td>\n",
       "      <td>89.587305</td>\n",
       "      <td>755.092999</td>\n",
       "      <td>82.477201</td>\n",
       "      <td>85.321243</td>\n",
       "      <td>...</td>\n",
       "      <td>5.688083</td>\n",
       "      <td>1161.790924</td>\n",
       "      <td>98.119429</td>\n",
       "      <td>21775.403204</td>\n",
       "      <td>52.614766</td>\n",
       "      <td>5039.641413</td>\n",
       "      <td>99.541450</td>\n",
       "      <td>142.202071</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 457 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CORO6        SULF1    LRRC37A3        CD22          HLA-B  \\\n",
       "HR1005   1.721086   888.080232  247.836344   34.421714   50704.906417   \n",
       "HR1006   4.264374   573.131894  147.547348  133.901350   63955.378653   \n",
       "HR1009   6.468368   681.766017  238.035953  227.686563   56408.052419   \n",
       "HR1010  14.335907  1333.993833  219.565727   43.762241   40282.388433   \n",
       "HR1012   5.688083   766.469165  191.972796   69.679015  207430.161649   \n",
       "\n",
       "            CSRNP3       EFHD1       CYB5R2     SLC2A6        SCG2  ...  \\\n",
       "HR1005  123.918172  211.693544   616.148688  25.816286  127.360343  ...   \n",
       "HR1006   12.793123  193.602589  1148.822412  51.172491   19.616121  ...   \n",
       "HR1009   87.969809  190.170027   922.389317  60.802662   42.691231  ...   \n",
       "HR1010   30.935377  418.004855   768.857305  38.480591   36.971549  ...   \n",
       "HR1012   45.504663   89.587305   755.092999  82.477201   85.321243  ...   \n",
       "\n",
       "            CADPS         RAC2        HSF4          CD24    PGLYRP4  \\\n",
       "HR1005   8.605429   304.632173   67.122343  36724.527108  30.979543   \n",
       "HR1006  14.498872   814.495474  150.105972  19742.346843  23.027621   \n",
       "HR1009  31.048168   459.254148  326.005761  27198.194942   9.055716   \n",
       "HR1010  27.917292   368.206443  390.087563  33152.161215  11.317821   \n",
       "HR1012   5.688083  1161.790924   98.119429  21775.403204  52.614766   \n",
       "\n",
       "                NTS        HBB       PHLDB2       Type      Batch  \n",
       "HR1005  9047.747634   0.000000   166.945315  asthmatic  GSE152004  \n",
       "HR1006  4201.261472   6.822999   894.665709  asthmatic  GSE152004  \n",
       "HR1009  4655.931488   7.762042   724.457247  asthmatic  GSE152004  \n",
       "HR1010  3876.730946  14.335907  1795.006407  asthmatic  GSE152004  \n",
       "HR1012  5039.641413  99.541450   142.202071  asthmatic  GSE152004  \n",
       "\n",
       "[5 rows x 457 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# prepare data for all common DEGs\n",
    "all_DEGs_data = all_common_genes_data[all_DEGs+[\"Type\", \"Batch\"]]\n",
    "all_DEGs_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get DCEGs for each dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "executionInfo": {
     "elapsed": 325,
     "status": "ok",
     "timestamp": 1710581276163,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "EidQ4bUkWZDq",
    "outputId": "4dfabe4e-16ba-48d2-c2a1-2a0f322ea314"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>colors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>100009676_at</td>\n",
       "      <td>brown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10000_at</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>10001_at</td>\n",
       "      <td>yellow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>10002_at</td>\n",
       "      <td>magenta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>10003_at</td>\n",
       "      <td>magenta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0       gene_id   colors\n",
       "0           1  100009676_at    brown\n",
       "1           2      10000_at    black\n",
       "2           3      10001_at   yellow\n",
       "3           4      10002_at  magenta\n",
       "4           5      10003_at  magenta"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE67472_modules_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 206
    },
    "executionInfo": {
     "elapsed": 13,
     "status": "ok",
     "timestamp": 1710581278835,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "WHswsIowWd13",
    "outputId": "c0d2d82a-4914-4407-b14f-150ebf9a8565"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Correlation</th>\n",
       "      <th>Pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MEgreenyellow</td>\n",
       "      <td>0.452231</td>\n",
       "      <td>1.279046e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MEturquoise</td>\n",
       "      <td>0.185027</td>\n",
       "      <td>5.880970e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MEgreen</td>\n",
       "      <td>0.567792</td>\n",
       "      <td>2.676138e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEmagenta</td>\n",
       "      <td>0.499118</td>\n",
       "      <td>5.977090e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MEblack</td>\n",
       "      <td>0.009015</td>\n",
       "      <td>9.272795e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0  Correlation        Pvalue\n",
       "0  MEgreenyellow     0.452231  1.279046e-06\n",
       "1    MEturquoise     0.185027  5.880970e-02\n",
       "2        MEgreen     0.567792  2.676138e-10\n",
       "3      MEmagenta     0.499118  5.977090e-08\n",
       "4        MEblack     0.009015  9.272795e-01"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE67472_r_and_Pvalue_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>colors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>LOC100009676</td>\n",
       "      <td>brown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>AKT3</td>\n",
       "      <td>black</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>MED6</td>\n",
       "      <td>yellow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>NR2E3</td>\n",
       "      <td>magenta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>NAALAD2</td>\n",
       "      <td>magenta</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0       gene_id   colors\n",
       "0           1  LOC100009676    brown\n",
       "1           2          AKT3    black\n",
       "2           3          MED6   yellow\n",
       "3           4         NR2E3  magenta\n",
       "4           5       NAALAD2  magenta"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## First annoatate Affimetrix genes\n",
    "GSE67472_modules_df[\"gene_id\"] = GSE67472_modules_df[\"gene_id\"].apply(lambda x: Affimetrix_map['ENTREZID_to_ALIAS'][x.split(\"_\")[0]] if x.split(\"_\")[0] in Affimetrix_map['ENTREZID_to_ALIAS'] else x)\n",
    "GSE67472_modules_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>colors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1007_PM_s_at</td>\n",
       "      <td>green</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>1053_PM_at</td>\n",
       "      <td>yellow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>117_PM_at</td>\n",
       "      <td>yellow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>121_PM_at</td>\n",
       "      <td>turquoise</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>1255_PM_g_at</td>\n",
       "      <td>turquoise</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0       gene_id     colors\n",
       "0           1  1007_PM_s_at      green\n",
       "1           2    1053_PM_at     yellow\n",
       "2           3     117_PM_at     yellow\n",
       "3           4     121_PM_at  turquoise\n",
       "4           5  1255_PM_g_at  turquoise"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GSE69683_modules_df[\"Unnamed: 0\"] = GSE69683_modules_df[\"Unnamed: 0\"].apply(lambda x: Affimetrix_map['AFF_PM_to_ALIAS'][x] if x in Affimetrix_map['AFF_PM_to_ALIAS'] else x)\n",
    "GSE69683_modules_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4472 4141 677 2352 9649\n"
     ]
    }
   ],
   "source": [
    "# get coexpressed genes per data\n",
    "ceg1 = set(get_coexpressed_genes(GSE152004_modules_df, GSE152004_r_and_Pvalue_df, r_thresh=0.1, Pvalue_thresh=0.05))\n",
    "ceg2 = set(get_coexpressed_genes(GSE201955_modules_df, GSE201955_r_and_Pvalue_df, r_thresh=0.1, Pvalue_thresh=0.05))\n",
    "ceg3 = set(get_coexpressed_genes(GSE58434_modules_df, GSE58434_r_and_Pvalue_df, r_thresh=0.1, Pvalue_thresh=0.05))\n",
    "ceg4 = set(get_coexpressed_genes(GSE67472_modules_df, GSE67472_r_and_Pvalue_df, r_thresh=0.1, Pvalue_thresh=0.05))\n",
    "ceg5 = set(get_coexpressed_genes(GSE69683_modules_df, GSE69683_r_and_Pvalue_df, r_thresh=0.1, Pvalue_thresh=0.05))\n",
    "\n",
    "print(len(ceg1), len(ceg2), len(ceg3), len(ceg4), len(ceg5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13 235 37 111 9\n"
     ]
    }
   ],
   "source": [
    "# get differentially co-expressed genes\n",
    "dceg1 = set(get_differentially_co_expressed_genes(deg1, ceg1))\n",
    "dceg2 = set(get_differentially_co_expressed_genes(deg2, ceg2))\n",
    "dceg3 = set(get_differentially_co_expressed_genes(deg3, ceg3))\n",
    "dceg4 = set(get_differentially_co_expressed_genes(deg4, ceg4))\n",
    "dceg5 = set(get_differentially_co_expressed_genes(deg5, ceg5))\n",
    "\n",
    "print(len(dceg1), len(dceg2), len(dceg3), len(dceg4), len(dceg5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All common DCEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    }
   ],
   "source": [
    "# get common differentially expressed genes\n",
    "all_common_DCEGs = dceg1.intersection(dceg2,dceg3,dceg4,dceg5)\n",
    "print(len(all_common_DCEGs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All DCEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "296\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['CORO6',\n",
       " 'NELL2',\n",
       " 'LRRC37A3',\n",
       " 'OST4',\n",
       " 'EPB41L2',\n",
       " 'IFI27',\n",
       " 'DOK1',\n",
       " 'FGFBP1',\n",
       " 'AGER',\n",
       " 'HLA-B']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all differentially expressed genes that have matching annotation\n",
    "all_DCEGs = list(dceg1.union(dceg2,dceg3,dceg4,dceg5).intersection(common_genes))\n",
    "print(len(all_DCEGs))\n",
    "all_DCEGs[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CORO6</th>\n",
       "      <th>NELL2</th>\n",
       "      <th>LRRC37A3</th>\n",
       "      <th>OST4</th>\n",
       "      <th>EPB41L2</th>\n",
       "      <th>IFI27</th>\n",
       "      <th>DOK1</th>\n",
       "      <th>FGFBP1</th>\n",
       "      <th>AGER</th>\n",
       "      <th>HLA-B</th>\n",
       "      <th>...</th>\n",
       "      <th>SCN2B</th>\n",
       "      <th>NTS</th>\n",
       "      <th>TK1</th>\n",
       "      <th>KCNJ1</th>\n",
       "      <th>HBB</th>\n",
       "      <th>C2orf72</th>\n",
       "      <th>SEMA3B</th>\n",
       "      <th>TPH1</th>\n",
       "      <th>Type</th>\n",
       "      <th>Batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HR1005</th>\n",
       "      <td>1.721086</td>\n",
       "      <td>10564.024154</td>\n",
       "      <td>247.836344</td>\n",
       "      <td>2316.581380</td>\n",
       "      <td>927.665203</td>\n",
       "      <td>3843.184414</td>\n",
       "      <td>409.618402</td>\n",
       "      <td>234.067658</td>\n",
       "      <td>24.095200</td>\n",
       "      <td>50704.906417</td>\n",
       "      <td>...</td>\n",
       "      <td>103.265143</td>\n",
       "      <td>9047.747634</td>\n",
       "      <td>99.822972</td>\n",
       "      <td>194.482686</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>115.312743</td>\n",
       "      <td>184.156172</td>\n",
       "      <td>6.884343</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1006</th>\n",
       "      <td>4.264374</td>\n",
       "      <td>2861.395095</td>\n",
       "      <td>147.547348</td>\n",
       "      <td>3466.936233</td>\n",
       "      <td>436.671919</td>\n",
       "      <td>3712.564187</td>\n",
       "      <td>1397.861866</td>\n",
       "      <td>1879.736152</td>\n",
       "      <td>28.144870</td>\n",
       "      <td>63955.378653</td>\n",
       "      <td>...</td>\n",
       "      <td>37.526493</td>\n",
       "      <td>4201.261472</td>\n",
       "      <td>156.076096</td>\n",
       "      <td>21.321871</td>\n",
       "      <td>6.822999</td>\n",
       "      <td>40.937992</td>\n",
       "      <td>295.094695</td>\n",
       "      <td>1.705750</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1009</th>\n",
       "      <td>6.468368</td>\n",
       "      <td>1701.180858</td>\n",
       "      <td>238.035953</td>\n",
       "      <td>3111.285143</td>\n",
       "      <td>360.934950</td>\n",
       "      <td>4421.776556</td>\n",
       "      <td>1492.899399</td>\n",
       "      <td>1099.622608</td>\n",
       "      <td>135.835734</td>\n",
       "      <td>56408.052419</td>\n",
       "      <td>...</td>\n",
       "      <td>84.088788</td>\n",
       "      <td>4655.931488</td>\n",
       "      <td>169.471249</td>\n",
       "      <td>65.977356</td>\n",
       "      <td>7.762042</td>\n",
       "      <td>40.103883</td>\n",
       "      <td>210.868806</td>\n",
       "      <td>11.643063</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1010</th>\n",
       "      <td>14.335907</td>\n",
       "      <td>1309.094627</td>\n",
       "      <td>219.565727</td>\n",
       "      <td>2765.320926</td>\n",
       "      <td>287.472653</td>\n",
       "      <td>2717.031557</td>\n",
       "      <td>760.557570</td>\n",
       "      <td>2761.548319</td>\n",
       "      <td>109.405603</td>\n",
       "      <td>40282.388433</td>\n",
       "      <td>...</td>\n",
       "      <td>18.863035</td>\n",
       "      <td>3876.730946</td>\n",
       "      <td>246.728497</td>\n",
       "      <td>52.061977</td>\n",
       "      <td>14.335907</td>\n",
       "      <td>9.808778</td>\n",
       "      <td>319.917073</td>\n",
       "      <td>3.772607</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HR1012</th>\n",
       "      <td>5.688083</td>\n",
       "      <td>1534.360351</td>\n",
       "      <td>191.972796</td>\n",
       "      <td>1770.415790</td>\n",
       "      <td>584.450514</td>\n",
       "      <td>32653.861671</td>\n",
       "      <td>581.606472</td>\n",
       "      <td>291.514247</td>\n",
       "      <td>44.082642</td>\n",
       "      <td>207430.161649</td>\n",
       "      <td>...</td>\n",
       "      <td>2.844041</td>\n",
       "      <td>5039.641413</td>\n",
       "      <td>250.275646</td>\n",
       "      <td>8.532124</td>\n",
       "      <td>99.541450</td>\n",
       "      <td>63.990932</td>\n",
       "      <td>109.495595</td>\n",
       "      <td>1.422021</td>\n",
       "      <td>asthmatic</td>\n",
       "      <td>GSE152004</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 298 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CORO6         NELL2    LRRC37A3         OST4     EPB41L2  \\\n",
       "HR1005   1.721086  10564.024154  247.836344  2316.581380  927.665203   \n",
       "HR1006   4.264374   2861.395095  147.547348  3466.936233  436.671919   \n",
       "HR1009   6.468368   1701.180858  238.035953  3111.285143  360.934950   \n",
       "HR1010  14.335907   1309.094627  219.565727  2765.320926  287.472653   \n",
       "HR1012   5.688083   1534.360351  191.972796  1770.415790  584.450514   \n",
       "\n",
       "               IFI27         DOK1       FGFBP1        AGER          HLA-B  \\\n",
       "HR1005   3843.184414   409.618402   234.067658   24.095200   50704.906417   \n",
       "HR1006   3712.564187  1397.861866  1879.736152   28.144870   63955.378653   \n",
       "HR1009   4421.776556  1492.899399  1099.622608  135.835734   56408.052419   \n",
       "HR1010   2717.031557   760.557570  2761.548319  109.405603   40282.388433   \n",
       "HR1012  32653.861671   581.606472   291.514247   44.082642  207430.161649   \n",
       "\n",
       "        ...       SCN2B          NTS         TK1       KCNJ1        HBB  \\\n",
       "HR1005  ...  103.265143  9047.747634   99.822972  194.482686   0.000000   \n",
       "HR1006  ...   37.526493  4201.261472  156.076096   21.321871   6.822999   \n",
       "HR1009  ...   84.088788  4655.931488  169.471249   65.977356   7.762042   \n",
       "HR1010  ...   18.863035  3876.730946  246.728497   52.061977  14.335907   \n",
       "HR1012  ...    2.844041  5039.641413  250.275646    8.532124  99.541450   \n",
       "\n",
       "           C2orf72      SEMA3B       TPH1       Type      Batch  \n",
       "HR1005  115.312743  184.156172   6.884343  asthmatic  GSE152004  \n",
       "HR1006   40.937992  295.094695   1.705750  asthmatic  GSE152004  \n",
       "HR1009   40.103883  210.868806  11.643063  asthmatic  GSE152004  \n",
       "HR1010    9.808778  319.917073   3.772607  asthmatic  GSE152004  \n",
       "HR1012   63.990932  109.495595   1.422021  asthmatic  GSE152004  \n",
       "\n",
       "[5 rows x 298 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# prepare data for all common DEGs\n",
    "all_DCEGs_data = all_common_genes_data[all_DCEGs+[\"Type\", \"Batch\"]]\n",
    "all_DCEGs_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "executionInfo": {
     "elapsed": 82766,
     "status": "ok",
     "timestamp": 1710581395089,
     "user": {
      "displayName": "Cyrille Mesue",
      "userId": "05628558888145213857"
     },
     "user_tz": -180
    },
    "id": "ZoQHDCyQkqgU"
   },
   "outputs": [],
   "source": [
    "### Save data\n",
    "batch = all_common_genes_data.Batch\n",
    "Type = all_common_genes_data.Type\n",
    "all_common_genes_data.to_csv(\"data2/all_common_genes_data.csv\", index = False)\n",
    "all_DEGs_data.to_csv(\"data2/all_DEGs_data.csv\", index = False)\n",
    "all_DCEGs_data.to_csv(\"data2/all_DCEGs_data.csv\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch Effect Correction Using PyCombat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 5 batches.\n",
      "Adjusting for 0 covariate(s) or covariate level(s).\n",
      "Standardizing Data across genes.\n",
      "Fitting L/S model and finding priors.\n",
      "Finding parametric adjustments.\n",
      "Adjusting the Data\n",
      "Found 5 batches.\n",
      "Adjusting for 0 covariate(s) or covariate level(s).\n",
      "Standardizing Data across genes.\n",
      "Fitting L/S model and finding priors.\n",
      "Finding parametric adjustments.\n",
      "Adjusting the Data\n",
      "Found 5 batches.\n",
      "Adjusting for 0 covariate(s) or covariate level(s).\n",
      "Standardizing Data across genes.\n",
      "Fitting L/S model and finding priors.\n",
      "Finding parametric adjustments.\n",
      "Adjusting the Data\n"
     ]
    }
   ],
   "source": [
    "# remove batch effect\n",
    "all_DEGs_data_without_batch = pycombat(all_DEGs_data.drop([\"Type\", \"Batch\"], axis =1).T,batch).T\n",
    "all_DCEGs_data_without_batch = pycombat(all_DCEGs_data.drop([\"Type\", \"Batch\"], axis =1).T,batch).T\n",
    "all_common_genes_data_without_batch = pycombat(all_common_genes_data.drop([\"Type\", \"Batch\"], axis =1).T,batch).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save corrected data\n",
    "all_common_genes_data_without_batch[\"Type\"] = Type\n",
    "all_DEGs_data_without_batch[\"Type\"] = Type\n",
    "all_DCEGs_data_without_batch[\"Type\"] = Type\n",
    "all_common_genes_data_without_batch.to_csv(\"data2/all_common_genes_data_without_batch.csv\", index = False)\n",
    "all_DEGs_data_without_batch.to_csv(\"data2/all_DEGs_data_without_batch.csv\", index = False)\n",
    "all_DCEGs_data_without_batch.to_csv(\"data2/all_DCEGs_data_without_batch.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyO4ZOyZbVvoUAaWn9A03xtp",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
